{
 "metadata": {
  "name": "",
  "signature": "sha256:b862b890e0756ea04af13f935807e2e3786b2357c02ba0f4d108160cd053192b"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Simulation of Whole-Cell Model of Translation in *E. coli* #\n",
      "\n",
      "Supplemental Information for\n",
      "\n",
      "**An Integrated Approach Reveals Regulatory Controls Underlying Bacterial Translation Elongation**, Cell, Volume 159, Issue 5, 2014\n",
      "\n",
      "<a href=\"mailto:asubram@fas.harvard.edu\">*Arvind R. Subramaniam*</a>, *Brian M. Zid*, and <a href=\"mailto:erin_oshea@harvard.edu\">*Erin K. O\u2019Shea*</a>"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Note: Finish pre-processing the sequencing data in the sequencing_data_analysis.ipynb script before running this simulation, since we use the ribosome profiling and rna seq measurements to infer translation initation rate and mrna copy number.\n",
      "\n",
      "#### Table of Contents ####\n",
      "\n",
      "* <a href=\"#input\"> Create input files for simulation</a>\n",
      "  * <a href=\"#trna\"> Create input files for trnas and codons</a>\n",
      "  * <a href=\"#mrna\"> Create input files for mrnas</a>\n",
      "  * <a href=\"#transgene\"> Create input files for transgenes</a>\n",
      "* <a href=\"#source\"> Source code for simulation</a>     \n",
      "* <a href=\"#compile\"> Compiling source</a> \n",
      "* <a href=\"#run\">Running the simulation</a>"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#### Define globals ####"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pandas # tabular data processing library.\n",
      "import cPickle as pickle # python native format storage library.\n",
      "import Bio.SeqIO # dna sequence input and output.\n",
      "from Bio.SeqUtils import CodonUsage # provides dictionary of codon names\n",
      "from Bio.SeqUtils import CodonUsageIndices # provides the Ecoli Codon \n",
      "                                            # Adaptation Index class.\n",
      "from Bio.SeqRecord import SeqRecord # to read .gb and .fasta files.\n",
      "import re # regular expression \n",
      "import numpy.random # for creating random numbers\n",
      "from Bio import Entrez # to retrieve records from NCBI\n",
      "import itertools # iterator tools\n",
      "from IPython.core.display import clear_output, HTML # for refreshing and displaying output\n",
      "\n",
      "Entrez.email = 'USE_YOUR_EMAIL_ID_HERE'\n",
      "\n",
      "# This file is taken from Shah et al. 2013 to keep \n",
      "# the same numerical index convention for each codon.\n",
      "geneticCode = pandas.read_table( 'seq_and_annotation/genetic.code.tsv' \n",
      "                                ).reset_index().set_index('codon') \n",
      "\n",
      "# Values are typically taken for 30 min doubling time whenever available.\n",
      "# Note that if you change the parameters here, \n",
      "# they will also have to changed in the ecoli_translation.c file below.\n",
      "fixed_parameters = { \n",
      "    'organism' : 'Escherichia coli',                    \n",
      "    'mean number of ribosomes' : 4.4e4, # ribosomes per cell. Bremer 2008\n",
      "    'fraction of mrna bound ribosomes' : 0.85, # Bremer 2008\n",
      "    'mean number of mrnas' : 4.0e6, # nucleotides per cell, Bremer 2008\n",
      "    'mean normal elongation rate': 66, # nucleotides per second, Bremer 2008\n",
      "    'kcat by km for trna incorporation':  2.0e7,  # M-1 s-1, Elf 2003\n",
      "    'mean number of trnas' : 4.08e5, # molecules per cell, Bremer 2008\n",
      "    'cell volume': 1.0e-18, # m-3, Bremer 2008\n",
      "    'ribosome footprint size': 10, # codons\n",
      "    'avogadro constant': 6.02e23,    \n",
      "    'max_aminoacylation_rate': 2.0e10,\n",
      "                    } "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<a id=\"input\"></a>\n",
      "### Create input files for simulation ###\n",
      "\n",
      "<a id=\"trna\"></a>\n",
      "#### Create input files for tRNAs and codons data ####\n",
      "There are four files for:\n",
      " * tRNA concentration\n",
      " * aminoacylation rate for each tRNA during rich growth\n",
      " * cognate codon-tRNA pairs\n",
      " * codon-trna interaction weights for cognate codon-tRNA pairs "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Read tRNA concentrations, anticodon, cognate codons \n",
      "# from Table 2 in Dong et al. J. Mol. Biol. 1996.\n",
      "lines = open( 'seq_and_annotation/Ecoli.trna.abundance.dong1996.txt'\n",
      "             ).read().splitlines()\n",
      "trnas = dict()\n",
      "#read every 6th line until last but one.\n",
      "for trnaindex, trna in enumerate( lines[0:-1:6] ): \n",
      "    trnas[ trna ] = dict()\n",
      "    trnas[ trna ]['anticodon'] = lines[trnaindex*6 + 1\n",
      "                                       ].strip().replace('U','T')\n",
      "    trnas[ trna ]['codons'] = [codon.strip().replace('U','T')\n",
      "                               for codon in lines[trnaindex*6 + 2\n",
      "                                                  ].split( ',' )]    \n",
      "    trnas[ trna ]['mean_concn'] = int( lines[trnaindex*6 + 4].split('(')[0] )\n",
      "    trnas[ trna ]['std_concn'] = int( lines[trnaindex*6 + 4\n",
      "                                            ].split('(')[1].split(')'\n",
      "                                                            )[0].strip() )\n",
      "    trnas[ trna ]['fraction'] = float( lines[trnaindex*6 + 5].strip() )    \n",
      "    trnas[ trna ]['aminoacid'] = trna[:3].lower()\n",
      "\n",
      "# Ile3 was not measured in Dong 1996. So assigning it to Ile1.    \n",
      "trnas[ 'Ile1' ]['codons'].append( 'ATA' ) \n",
      "# Leu3 was found to read CTT in S\u00f8rensen et al. J. Mol. Biol. 2005.\n",
      "trnas[ 'Leu3' ]['codons'].append( 'CTT' )\n",
      "\n",
      "# Calculate the factors for codon-tRNA interaction \n",
      "# that multiply the kcat/KM for codon-tRNA reading:\n",
      "# w = 1 for Watson-Crick base pairs,\n",
      "# w = 0.64 for purine-pyrimidine mismatch,\n",
      "# w = 0.61 for purine-purine mismatch. (Weights based on Shah 2013).\n",
      "\n",
      "# If two trnas have the same anticodon \n",
      "# they are considered as a single trna species in the simulation.\n",
      "concentrations = dict() \n",
      "cognate_pairings = dict()\n",
      "weights = dict()\n",
      "# For Fig. S1A, weighted sum of all cognate trna concentrations for each codon \n",
      "trna_abundance = dict() \n",
      "\n",
      "for trna in trnas:    \n",
      "    cognatecodons = trnas[trna]['codons']\n",
      "    if trnas[trna]['anticodon'] not in concentrations:\n",
      "        concentrations[trnas[trna]['anticodon']] = trnas[trna]['mean_concn']\n",
      "    else:\n",
      "        concentrations[trnas[trna]['anticodon']] += trnas[trna]['mean_concn']\n",
      "        \n",
      "    for codon in cognatecodons: \n",
      "        # pairing at the wobble position.\n",
      "        pairing = codon[-1:] + trnas[trna]['anticodon'][:1] \n",
      "        watson_crick_pairs = ['GC','AT','TA','CG']\n",
      "        pur_pyr_pairs = ['GT','TG','CA']\n",
      "        pur_pur_pairs = ['TT','AA']\n",
      "        if pairing in watson_crick_pairs:\n",
      "            weight = 1\n",
      "        elif pairing in pur_pyr_pairs:\n",
      "            weight = 0.64\n",
      "        else:\n",
      "            weight = 0.61\n",
      "            \n",
      "        # Cognate tRNA assignment below based on data in S\u00f8rensen et al. J. Mol. Biol. 2005.    \n",
      "        if codon in ['CTT'] and trnas[trna]['anticodon'] in ['TAG']:\n",
      "            weight = 1.0 \n",
      "            \n",
      "        if codon not in cognate_pairings:\n",
      "            cognate_pairings[ codon ] = [ trnas[trna]['anticodon'] ]\n",
      "            weights[ codon ] = [ weight ]            \n",
      "        elif trnas[trna]['anticodon'] not in cognate_pairings[ codon ]:\n",
      "            cognate_pairings[ codon ].append( trnas[trna]['anticodon'] )\n",
      "            weights[ codon ].append( weight )\n",
      "            \n",
      "        if codon.replace('U','T') not in trna_abundance.keys():\n",
      "            trna_abundance[ codon.replace('U','T') ] = int( \n",
      "                        trnas[trna]['mean_concn'] * weight )\n",
      "        else:\n",
      "            trna_abundance[ codon.replace('U','T') ] = int(\n",
      "                        trna_abundance[ codon.replace('U','T') ] \n",
      "                                + trnas[trna]['mean_concn'] * weight )\n",
      "            \n",
      "codon_parameters = pandas.DataFrame.from_dict( \n",
      "                            { 'trna_weighted_concn' : trna_abundance } )\n",
      "codon_parameters[['trna_weighted_concn']].to_csv(\n",
      "    'seq_and_annotation/Ecoli.trna.concn.for.codon.csv', index='codon')\n",
      "\n",
      "#trnaindices are sorted alphabetically.\n",
      "\n",
      "# concentrations file.            \n",
      "outputFile = open( \"seq_and_annotation/Ecoli.trna.concentrations.tsv\"\n",
      "                  , \"write\" )\n",
      "outputFile.write( 'trnaindex\\tanticodon\\tmolpercell\\taminoacid\\n' )\n",
      "# trnaindex 0 is a dummy trna without any anticodon.\n",
      "outputFile.write( '0\\tNNN\\t0\\txxx\\n' ) \n",
      "trnaindices = dict()\n",
      "for trnaindex, anticodon in enumerate( sorted(concentrations) ):\n",
      "    aminoacid = [ trnas[trna]['aminoacid'] \n",
      "                 for trna in trnas \n",
      "                 if trnas[trna]['anticodon'] == anticodon ][0]\n",
      "    trnaindices[ anticodon ] = trnaindex + 1\n",
      "    outputFile.write( str(trnaindex + 1) + '\\t' \n",
      "                     + anticodon + '\\t' + '%d'%concentrations[anticodon] \n",
      "                     + '\\t' + aminoacid + '\\n' )    \n",
      "outputFile.close()  \n",
      "\n",
      "# aminoacylation rates file.\n",
      "outputFile = open( \"seq_and_annotation/Ecoli.trna.aminoacylation.rate.tsv\"\n",
      "                  , \"write\" )\n",
      "outputFile.write( 'trnaindex\\tanticodon\\tkcatBykm\\n' )\n",
      "outputFile.write( '0\\tNNN\\t0\\n' )\n",
      "trnaindices = dict()\n",
      "for trnaindex, anticodon in enumerate( sorted(concentrations) ):\n",
      "    trnaindices[ anticodon ] = trnaindex + 1\n",
      "    outputFile.write( str(trnaindex + 1) + '\\t'\n",
      "                     + anticodon + '\\t' \n",
      "                     + '%d'% fixed_parameters['max_aminoacylation_rate'] \n",
      "                     + '\\n' )    \n",
      "outputFile.close() \n",
      "\n",
      "# cognate pairings file.\n",
      "outputFile = open( \"seq_and_annotation/Ecoli.cognate.pairs.tsv\", \"write\" )\n",
      "outputFile.write( 'codonindex\\tcodon\\ttrnaindex1\\ttrnaindex2\\n' )\n",
      "for codon in geneticCode.index:\n",
      "    if len( cognate_pairings[codon] ) == 1:    \n",
      "        outputFile.write( str(geneticCode.ix[codon]['index']) \n",
      "                     + '\\t' + codon + '\\t' \n",
      "                     + '%d'%trnaindices[cognate_pairings[codon][0]] \n",
      "                     + '\\t0\\n' )\n",
      "    elif len( cognate_pairings[codon] ) == 2:    \n",
      "        outputFile.write( str(geneticCode.ix[codon]['index']) \n",
      "                     + '\\t' + codon                           \n",
      "                     + '\\t' + '%d'%trnaindices[cognate_pairings[codon][0]] \n",
      "                     + '\\t' + '%d'%trnaindices[cognate_pairings[codon][1]] \n",
      "                     + '\\n' )\n",
      "outputFile.close() \n",
      "\n",
      "# cognate weights file.\n",
      "outputFile = open( \"seq_and_annotation/Ecoli.cognate.weights.tsv\", \"write\" )\n",
      "outputFile.write( 'codonindex\\tcodon\\ttrnaweight1\\ttrnaweight2\\n' )\n",
      "for codon in geneticCode.index:\n",
      "    if len( weights[codon] ) == 1:    \n",
      "        outputFile.write( str(geneticCode.ix[codon]['index']) \n",
      "                         + '\\t' + codon + '\\t' \n",
      "                         + '%0.2f'%weights[codon][0] + '\\t0.00\\n' )\n",
      "    elif len( weights[codon] ) == 2:    \n",
      "        outputFile.write( str(geneticCode.ix[codon]['index'])\n",
      "                         + '\\t' + codon + \n",
      "                         '\\t' + '%0.2f'%weights[codon][0] \n",
      "                         + '\\t' + '%0.2f'%weights[codon][1] + '\\n' )\n",
      "outputFile.close() "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$1 / k^{el} n^{at}$ values for 61 codons (in milliseconds)"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "temp = codon_parameters[['trna_weighted_concn']].apply( lambda trna_concentration: \n",
      "                                                       fixed_parameters['avogadro constant']\n",
      "                                                       / fixed_parameters['kcat by km for trna incorporation'] \n",
      "                                                          * fixed_parameters['cell volume']  * 1e3 # m-3 to liters\n",
      "                                                          / fixed_parameters['mean number of trnas'] \n",
      "                                                          / trna_concentration * 1e3 # seconds to milliseconds\n",
      "                                                          * sum( concentrations.values() ) )\n",
      "HTML( temp.sort(columns=['trna_weighted_concn']).to_html() )"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>trna_weighted_concn</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>CTG</th>\n",
        "      <td>  0.965201</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CGT</th>\n",
        "      <td>  0.994450</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GAA</th>\n",
        "      <td>  1.001829</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GGC</th>\n",
        "      <td>  1.084108</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GTA</th>\n",
        "      <td>  1.230632</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ATC</th>\n",
        "      <td>  1.360284</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GCA</th>\n",
        "      <td>  1.454039</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GTT</th>\n",
        "      <td>  1.499723</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CGC</th>\n",
        "      <td>  1.553971</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GAG</th>\n",
        "      <td>  1.565814</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CGA</th>\n",
        "      <td>  1.630651</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GGT</th>\n",
        "      <td>  1.694380</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ATG</th>\n",
        "      <td>  1.795451</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TTG</th>\n",
        "      <td>  1.837335</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GTG</th>\n",
        "      <td>  1.923332</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GAC</th>\n",
        "      <td>  1.972298</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ATT</th>\n",
        "      <td>  2.125788</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GGA</th>\n",
        "      <td>  2.211337</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ATA</th>\n",
        "      <td>  2.230121</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GCG</th>\n",
        "      <td>  2.271936</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TAC</th>\n",
        "      <td>  2.327895</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GCT</th>\n",
        "      <td>  2.384272</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>AAA</th>\n",
        "      <td>  2.456147</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TGC</th>\n",
        "      <td>  2.977710</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GAT</th>\n",
        "      <td>  3.082600</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>AGC</th>\n",
        "      <td>  3.356269</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GGG</th>\n",
        "      <td>  3.456932</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ACT</th>\n",
        "      <td>  3.569204</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TAT</th>\n",
        "      <td>  3.637895</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TCA</th>\n",
        "      <td>  3.646317</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TCT</th>\n",
        "      <td>  3.697673</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CCG</th>\n",
        "      <td>  3.718038</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CTT</th>\n",
        "      <td>  3.723898</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GTC</th>\n",
        "      <td>  3.735673</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>AAG</th>\n",
        "      <td>  3.838852</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ACC</th>\n",
        "      <td>  3.941306</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>AAC</th>\n",
        "      <td>  3.961128</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TCG</th>\n",
        "      <td>  4.028667</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ACG</th>\n",
        "      <td>  4.193102</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TTC</th>\n",
        "      <td>  4.557017</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TTA</th>\n",
        "      <td>  4.583537</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TGT</th>\n",
        "      <td>  4.655789</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CTC</th>\n",
        "      <td>  5.011269</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TGG</th>\n",
        "      <td>  5.011269</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ACA</th>\n",
        "      <td>  5.158981</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>AGT</th>\n",
        "      <td>  5.244868</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CAG</th>\n",
        "      <td>  5.363934</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>AGA</th>\n",
        "      <td>  5.450549</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CCT</th>\n",
        "      <td>  5.805438</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TCC</th>\n",
        "      <td>  6.185375</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CAA</th>\n",
        "      <td>  6.185375</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>AAT</th>\n",
        "      <td>  6.193481</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CCC</th>\n",
        "      <td>  6.563370</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CTA</th>\n",
        "      <td>  7.095535</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>TTT</th>\n",
        "      <td>  7.127641</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CAC</th>\n",
        "      <td>  7.395346</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CGG</th>\n",
        "      <td>  7.395346</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GCC</th>\n",
        "      <td>  7.659038</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CCA</th>\n",
        "      <td>  8.133608</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>AGG</th>\n",
        "      <td> 11.251491</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CAT</th>\n",
        "      <td> 11.582417</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 4,
       "text": [
        "<IPython.core.display.HTML at 0xb21513ac>"
       ]
      }
     ],
     "prompt_number": 4
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#### Modify the codon-tRNA interaction rates to use the exact measurement values from Sorensen et al. J. Mol. Biol. 2005. ####"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Read tRNA concentrations, anticodon, cognate codons \n",
      "# from Table 2 in Dong et al. J. Mol. Biol. 1996.\n",
      "lines = open( 'seq_and_annotation/Ecoli.trna.abundance.dong1996.txt'\n",
      "             ).read().splitlines()\n",
      "trnas = dict()\n",
      "#read every 6th line until last but one.\n",
      "for trnaindex, trna in enumerate( lines[0:-1:6] ): \n",
      "    trnas[ trna ] = dict()\n",
      "    trnas[ trna ]['anticodon'] = lines[trnaindex*6 + 1\n",
      "                                       ].strip().replace('U','T')\n",
      "    trnas[ trna ]['codons'] = [codon.strip().replace('U','T')\n",
      "                               for codon in lines[trnaindex*6 + 2\n",
      "                                                  ].split( ',' )]    \n",
      "    trnas[ trna ]['mean_concn'] = int( lines[trnaindex*6 + 4].split('(')[0] )\n",
      "    trnas[ trna ]['std_concn'] = int( lines[trnaindex*6 + 4\n",
      "                                            ].split('(')[1].split(')'\n",
      "                                                            )[0].strip() )\n",
      "    trnas[ trna ]['fraction'] = float( lines[trnaindex*6 + 5].strip() )    \n",
      "    trnas[ trna ]['aminoacid'] = trna[:3].lower()\n",
      "\n",
      "# Ile3 was not measured in Dong 1996. So assigning it to Ile1.    \n",
      "trnas[ 'Ile1' ]['codons'].append( 'ATA' ) \n",
      "# Leu3 was found to read CTT in S\u00f8rensen et al. J. Mol. Biol. 2005.\n",
      "trnas[ 'Leu3' ]['codons'].append( 'CTT' )\n",
      "\n",
      "# Calculate the factors for codon-tRNA interaction \n",
      "# that multiply the kcat/KM for codon-tRNA reading:\n",
      "# w = 1 for Watson-Crick base pairs,\n",
      "# w = 0.64 for purine-pyrimidine mismatch,\n",
      "# w = 0.61 for purine-purine mismatch. (Weights based on Shah 2013).\n",
      "\n",
      "# If two trnas have the same anticodon \n",
      "# they are considered as a single trna species in the simulation.\n",
      "concentrations = dict() \n",
      "cognate_pairings = dict()\n",
      "weights = dict()\n",
      "# For Fig. S1A, weighted sum of all cognate trna concentrations for each codon \n",
      "trna_abundance = dict() \n",
      "\n",
      "for trna in trnas:    \n",
      "    cognatecodons = trnas[trna]['codons']\n",
      "    if trnas[trna]['anticodon'] not in concentrations:\n",
      "        concentrations[trnas[trna]['anticodon']] = trnas[trna]['mean_concn']\n",
      "    else:\n",
      "        concentrations[trnas[trna]['anticodon']] += trnas[trna]['mean_concn']\n",
      "        \n",
      "    for codon in cognatecodons: \n",
      "        # pairing at the wobble position.\n",
      "        pairing = codon[-1:] + trnas[trna]['anticodon'][:1] \n",
      "        watson_crick_pairs = ['GC','AT','TA','CG']\n",
      "        pur_pyr_pairs = ['GT','TG','CA']\n",
      "        pur_pur_pairs = ['TT','AA']\n",
      "        if pairing in watson_crick_pairs:\n",
      "            weight = 1\n",
      "        elif pairing in pur_pyr_pairs:\n",
      "            weight = 0.64\n",
      "        else:\n",
      "            weight = 0.61\n",
      "            \n",
      "        # Based on data in S\u00f8rensen et al. J. Mol. Biol. 2005.\n",
      "        # Default weight for all other codons is 1 (kel = 2x10^7 M-1s-1).\n",
      "        if codon in ['CTG']:\n",
      "            if trnas[trna]['anticodon'] in ['CAG']:\n",
      "                weight = 18.2 / 20.0 \n",
      "            if trnas[trna]['anticodon'] in ['TAG']:\n",
      "                weight = 9.6 / 20.0 \n",
      "        if codon in ['CTC']:\n",
      "            if trnas[trna]['anticodon'] in ['GAG']:\n",
      "                weight = 30.3 / 20.0 \n",
      "        if codon in ['CTT']:\n",
      "            if trnas[trna]['anticodon'] in ['TAG']:\n",
      "                weight = 11.4 / 20.0 \n",
      "            if trnas[trna]['anticodon'] in ['GAG']:\n",
      "                weight = 32.8 / 20.0 \n",
      "        if codon in ['CTA']:\n",
      "            if trnas[trna]['anticodon'] in ['TAG']:\n",
      "                weight = 11.6 / 20.0\n",
      "            \n",
      "        if codon not in cognate_pairings:\n",
      "            cognate_pairings[ codon ] = [ trnas[trna]['anticodon'] ]\n",
      "            weights[ codon ] = [ weight ]            \n",
      "        elif trnas[trna]['anticodon'] not in cognate_pairings[ codon ]:\n",
      "            cognate_pairings[ codon ].append( trnas[trna]['anticodon'] )\n",
      "            weights[ codon ].append( weight )\n",
      "            \n",
      "        if codon.replace('U','T') not in trna_abundance.keys():\n",
      "            trna_abundance[ codon.replace('U','T') ] = int( \n",
      "                        trnas[trna]['mean_concn'] * weight )\n",
      "        else:\n",
      "            trna_abundance[ codon.replace('U','T') ] = int(\n",
      "                        trna_abundance[ codon.replace('U','T') ] \n",
      "                                + trnas[trna]['mean_concn'] * weight )\n",
      "\n",
      "# cognate weights file.\n",
      "outputFile = open( \"seq_and_annotation/Ecoli.cognate.weights.sorensen.tsv\", \"write\" )\n",
      "outputFile.write( 'codonindex\\tcodon\\ttrnaweight1\\ttrnaweight2\\n' )\n",
      "for codon in geneticCode.index:\n",
      "    if len( weights[codon] ) == 1:    \n",
      "        outputFile.write( str(geneticCode.ix[codon]['index']) \n",
      "                         + '\\t' + codon + '\\t' \n",
      "                         + '%0.2f'%weights[codon][0] + '\\t0.00\\n' )\n",
      "    elif len( weights[codon] ) == 2:    \n",
      "        outputFile.write( str(geneticCode.ix[codon]['index'])\n",
      "                         + '\\t' + codon + \n",
      "                         '\\t' + '%0.2f'%weights[codon][0] \n",
      "                         + '\\t' + '%0.2f'%weights[codon][1] + '\\n' )\n",
      "outputFile.close() "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 33
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<a id=\"mrna\"></a>\n",
      "#### Parse *E. coli* genome sequence and coding sequence annotations to create a fasta file of all coding sequences ####"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# .fna and .ptt files below downloaded from NCBI on 19th Feb 2014\n",
      "genome = Bio.SeqIO.read( 'seq_and_annotation/NC_000913.fna', 'fasta' )\n",
      "geneTable = pandas.read_table( 'seq_and_annotation/NC_000913.ptt',\n",
      "                              skiprows=2 )\n",
      "# Parse start coordinate of each CDS.\n",
      "geneTable['Start'] = geneTable['Location'].apply( \n",
      "                    lambda x: numpy.int32( x.split('.')[0] ) ) \n",
      "# Part end coordinate of each CDS.\n",
      "geneTable['End'] = geneTable['Location'].apply( \n",
      "                    lambda x: numpy.int32( x.split('.')[-1] ) )\n",
      "\n",
      "outputFile = open( 'seq_and_annotation/Ecoli.orfs.fasta', 'write' )\n",
      "genes = list()\n",
      "for row in geneTable.iterrows():\n",
      "    gene = row[1]['Gene']\n",
      "    if gene in genes:\n",
      "        gene = gene + row[1]['Synonym']\n",
      "    # The length of these gene sequences is not a multiple of 3, \n",
      "    # or contains stop codons, and are excluded from simulation.\n",
      "    if re.search( 'prfB|nrdF|fdnG|fdoG|fdhF', gene ): \n",
      "        continue \n",
      "    #1-based .ptt file indexing to 0-based python indexing.\n",
      "    start = row[1]['Start'] - 1 \n",
      "    # no -1 necessary since end-point is included in .ptt but not in python.\n",
      "    stop = row[1]['End'] \n",
      "    strand = row[1]['Strand']\n",
      "    seq = genome[ start:stop ].seq\n",
      "    # reverse_complement if on reverse strand.\n",
      "    if strand == '-': \n",
      "        seq = seq.reverse_complement()        \n",
      "    outputFile.write( '>' + gene + '\\n' + str(seq) + '\\n' ) \n",
      "    genes.append( gene )\n",
      "outputFile.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 40
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#### Calculate initiation rates and mRNA copy numbers for each coding sequence, and write it to an outputfile that also includes the numeric codon sequence. ####"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "samples = ['wt_yesleu_mono', 'wt_yesleu_total', 'wt_noleu_mono', 'wt_noleu_total']\n",
      "\n",
      "df = list()\n",
      "\n",
      "for sample in samples:    \n",
      "    temp = pandas.DataFrame.from_dict(  pickle.load( open( 'output_files/sequencing_output/geneStats_'+sample+'.pickled', 'rb' ) ), orient='index' )                        \n",
      "    df.append( temp.drop( ['meanReadDensity'], axis=1 ).rename( columns = {'totalReads': 'counts_' + sample} ) )        \n",
      "    \n",
      "geneTable = pandas.read_table( 'seq_and_annotation/NC_000913.ptt', skiprows=2 )\n",
      "geneTable['Start'] = geneTable['Location'].apply( lambda x: numpy.int32( x.split('.')[0] ) ) # Parse start coordinate of each CDS.\n",
      "geneTable['End'] = geneTable['Location'].apply( lambda x: numpy.int32( x.split('.')[-1] ) ) #Part end coordinate of each CDS.\n",
      "\n",
      "genes = dict()\n",
      "for row in geneTable.iterrows():\n",
      "    gene = row[1]['Gene']\n",
      "    if gene in genes:\n",
      "        gene = gene + row[1]['Synonym']\n",
      "    # The length of these gene sequences is not a multiple of 3, \n",
      "    # or contains stop codons, and are excluded from simulation.\n",
      "    if re.search( 'prfB|nrdF|fdnG|fdoG|fdhF', gene ): \n",
      "        continue \n",
      "    genes[ gene ] = {\n",
      "                     'function': row[1]['Product'],\n",
      "                     'strand':  row[1]['Strand'],\n",
      "                     'start': row[1]['Start'] - 1,\n",
      "                     'stop': row[1]['End'],\n",
      "                     'length': row[1]['End'] - row[1]['Start'] + 1 - 3 # exclude stop codon in length\n",
      "                     }\n",
      "df.append( pandas.DataFrame.from_dict(genes, orient='index') )\n",
      "        \n",
      "data = pandas.concat( df, join='inner', axis=1 )\n",
      "\n",
      "for col in data:\n",
      "    if col.startswith( 'counts' ):\n",
      "        data[col] = data[col].apply( numpy.float )        \n",
      "        data[ col.replace('counts','lengthnorm') ] =  numpy.divide( data[ col ], data[ 'length' ] )\n",
      "        if col.endswith('total'):\n",
      "            cellNormalization = sum( data[ col ] ) / fixed_parameters['mean number of mrnas']            \n",
      "        elif col.endswith('mono'):\n",
      "            cellNormalization = sum( data[ col ] ) / ( fixed_parameters['mean number of ribosomes']\n",
      "                                                    * fixed_parameters['fraction of mrna bound ribosomes'] )\n",
      "        \n",
      "        data[ col.replace('counts','cellnorm') ] = data[ col ].apply( lambda x: x  / cellNormalization )        \n",
      "        \n",
      "initiationRateList = []\n",
      "for row in data.iterrows():\n",
      "    # exclude genes that do not pass this read density threshold.\n",
      "    if not row[1]['lengthnorm_wt_noleu_mono'] > 1.0 / 3.0:\n",
      "        continue\n",
      "    # if genes get > 10 reads in ribosome profiling and rna seq, infer initiation rate. If not, assign\n",
      "    # the median initiation rate.\n",
      "    if row[1]['counts_wt_yesleu_mono'] > 10 and row[1]['counts_wt_yesleu_total'] > 10:\n",
      "        initiationRateList.append( row[1]['cellnorm_wt_yesleu_mono']\n",
      "                                  / row[1]['cellnorm_wt_yesleu_total']\n",
      "                                  * fixed_parameters['mean normal elongation rate'] )\n",
      "\n",
      "medianInitiationRate = numpy.median( initiationRateList )\n",
      "\n",
      "outputFile = open( 'seq_and_annotation/init_rates_and_mrna_numbers.csv', 'write' )\n",
      "outputFile.write( 'gene\\tinitiation rate (sec)\\tmRNA copy number (per cell)\\tgene class\\n' )\n",
      "totalNumberOfMrnas = 0\n",
      "geneCount = 0\n",
      "for row in data.iterrows():\n",
      "    # exclude genes that do not pass this read density threshold.\n",
      "    if not row[1]['lengthnorm_wt_noleu_mono'] > 1.0 / 3.0:\n",
      "        continue\n",
      "    if row[1]['counts_wt_yesleu_mono'] > 10 and row[1]['counts_wt_yesleu_total'] > 10:\n",
      "        initiationRate = ( row[1]['cellnorm_wt_yesleu_mono'] \n",
      "                            / row[1]['cellnorm_wt_yesleu_total'] \n",
      "                            * fixed_parameters['mean normal elongation rate'] )\n",
      "    else:\n",
      "        initiationRate = medianInitiationRate\n",
      "        \n",
      "    gene = row[0]\n",
      "\n",
      "    # If inferred mRNA copy number is less than 1, change it to 1.\n",
      "    if row[1]['cellnorm_wt_yesleu_total']/ row[1]['length'] > 1.0:\n",
      "        mrnaCopyNumber = numpy.ceil( row[1]['cellnorm_wt_yesleu_total'] / row[1]['length'] )\n",
      "    else:\n",
      "        mrnaCopyNumber = 1\n",
      "        \n",
      "    outputFile.write( '%s\\t%0.6f\\t%d\\n'%(row[0],initiationRate,mrnaCopyNumber) )\n",
      "    geneCount += 1\n",
      "    totalNumberOfMrnas += mrnaCopyNumber\n",
      "\n",
      "outputFile.close()\n",
      "\n",
      "\n",
      "inputFile = 'seq_and_annotation/init_rates_and_mrna_numbers.csv'\n",
      "initRateAndMrnaCopyNumbers = pandas.read_table( inputFile, index_col=0 )\n",
      "outputFile = open( 'seq_and_annotation/Ecoli.input.seq', 'write' )\n",
      "record_count = 0\n",
      "for record in Bio.SeqIO.parse( 'seq_and_annotation/Ecoli.orfs.fasta', 'fasta' ):\n",
      "    if record.id in initRateAndMrnaCopyNumbers.index:\n",
      "        outputFile.write( '%0.7f'%initRateAndMrnaCopyNumbers.ix[record.id]['initiation rate (sec)'] + ' ' +\n",
      "                              str(int(initRateAndMrnaCopyNumbers.ix[record.id]['mRNA copy number (per cell)'])) + ' ' )\n",
      "        for loop in range( 0, len(record)-6, 3 ):\n",
      "            outputFile.write( str( geneticCode.ix[ str(record.seq[loop:loop+3] )]['index'] ) + ' ' )\n",
      "        outputFile.write( str( geneticCode.ix[ str(record.seq[-6:-3] )]['index'] ) + '\\n' )            \n",
      "        record_count += 1\n",
      "outputFile.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<a id=\"transgene\"></a>\n",
      "#### Create transgene sequences with different CAI ####"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sharpEcoliIndex = CodonUsageIndices.SharpEcoliIndex\n",
      "cognateCodons = CodonUsage.SynonymousCodons\n",
      "sharpEcoliCai = CodonUsage.CodonAdaptationIndex()\n",
      "\n",
      "# function starts with an input sequence and lowers cai by random synonymous mutations \n",
      "# till the desired cai target is reached.\n",
      "def getSeqForTargetCai( inputSeq, targetCai ):\n",
      "    currentSeq = inputSeq\n",
      "    currentCai = sharpEcoliCai.cai_for_gene( currentSeq )\n",
      "    seqLength = len(inputSeq) / 3\n",
      "    numberOfIterations = 0\n",
      "    while currentCai > targetCai and numberOfIterations < 10000:\n",
      "        randPos = numpy.random.randint(0, seqLength )\n",
      "        oldCodon = currentSeq[3*randPos:3*(randPos+1)]\n",
      "        oldCodonCai = sharpEcoliIndex[ str(oldCodon) ]\n",
      "        aa = Bio.SeqUtils.seq3( Bio.Seq.Seq(oldCodon).translate() ).upper()\n",
      "        newCodon = numpy.random.choice( cognateCodons[ aa ] )\n",
      "        newCodonCai = sharpEcoliIndex[ newCodon ]\n",
      "        if (newCodonCai < oldCodonCai):\n",
      "            currentSeq = currentSeq[:3*randPos] + newCodon + currentSeq[3*(randPos+1):]            \n",
      "        currentCai = sharpEcoliCai.cai_for_gene( currentSeq )\n",
      "        numberOfIterations += 1\n",
      "        if numberOfIterations%100 == 0:\n",
      "            clear_output()\n",
      "            print protein, targetCai, currentCai, numberOfIterations\n",
      "            sys.stdout.flush()\n",
      "    return currentSeq\n",
      "\n",
      "\n",
      "ncbiproteins = {'insulin' : 'NP_000198.1',}\n",
      "\n",
      "transgeneProteins = dict()\n",
      "\n",
      "for protein in ncbiproteins:\n",
      "    temp = Entrez.efetch( db='protein', \n",
      "                            id=ncbiproteins[protein], \n",
      "                            rettype=\"fasta\", \n",
      "                            retmode=\"text\")\n",
      "    transgeneProteins[protein] = str( Bio.SeqIO.read( temp, 'fasta' ).seq )\n",
      "    \n",
      "# phi29 protein sequence from http://www.uniprot.org/uniprot/P03680\n",
      "transgeneProteins[ 'phi29pol' ] = str( Bio.SeqIO.read( 'seq_and_annotation/phi29pol.fa', 'fasta' ).seq )\n",
      "# yfp0 nucleotide sequence from Subramaniam et al. PNAS 2013\n",
      "transgeneProteins[ 'yfp' ] = str( Bio.SeqIO.read( 'seq_and_annotation/yfp0.fasta', 'fasta' ).seq.translate()[:-1] )\n",
      "\n",
      "for caiValue in numpy.arange(0.1,1,0.1): \n",
      "    seqrecords = list()\n",
      "    for protein in transgeneProteins:\n",
      "        #find the nucleotide sequence with cai value of 1.\n",
      "        maxCaiVariant = list()\n",
      "        for aa in transgeneProteins[protein]:\n",
      "            for codon in cognateCodons[ Bio.SeqUtils.seq3(aa).upper() ]:\n",
      "                if sharpEcoliIndex[ codon ] == 1:\n",
      "                    break\n",
      "            maxCaiVariant.append(codon)\n",
      "        maxCaiVariant = ''.join(maxCaiVariant)\n",
      "        \n",
      "        #create 10 random variants with the required cai value.\n",
      "        for variant in range(10):\n",
      "            newSeq = getSeqForTargetCai( maxCaiVariant, caiValue )\n",
      "            fasta_id = protein + \"_cai_\" + \"%0.1f\"%caiValue + \"_variant_\" + str(variant+1)            \n",
      "            seqrecords.append( SeqRecord( Bio.Seq.Seq( newSeq), id=fasta_id, description='') )\n",
      "            \n",
      "    outputFile = \"seq_and_annotation/Ecoli.transgene.cai\" + \"%0.1f\"%caiValue + \".fasta\"\n",
      "    Bio.SeqIO.write( seqrecords, outputFile, \"fasta\" )\n",
      "clear_output()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 42
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#### Create input seq files with transgene sequences ####"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# These renormalized values for mrnaFraction account for the fact that whenever the copy number of an mRNA\n",
      "# falls below 1 due to transgene expression, it is set to 1. This changes the relative\n",
      "# copy numbers slightly, but keeps all the native mRNAs in the simulation.\n",
      "mrnaFractionDict = {1:2.5, 10:15, 50:61}\n",
      "\n",
      "for actualFraction,mrnaFraction in mrnaFractionDict.items():\n",
      "    for cai in numpy.arange(0.1, 1, 0.1):\n",
      "        outputFile = open( \"seq_and_annotation/Ecoli.input.transgenefrac\"\n",
      "                          + str(actualFraction) + \".cai\" + str(cai) + \".seq\", \"write\" )     \n",
      "        totalNumberOfOldMrnas = 0\n",
      "        totalNumberOfNewMrnas = 0\n",
      "        initiationRateList = list()\n",
      "        for line in open( \"seq_and_annotation/Ecoli.input.seq\" ).readlines():\n",
      "            oldLine = line.split()\n",
      "            oldMrnaNumber = int( oldLine[1] )\n",
      "            initiationRateList.append( float( oldLine[0] ) )\n",
      "            totalNumberOfOldMrnas += oldMrnaNumber\n",
      "            newMrnaNumber = int( ceil( oldMrnaNumber * (100 - mrnaFraction) / 100.0 ) )\n",
      "            totalNumberOfNewMrnas += newMrnaNumber\n",
      "            newLine = \" \".join( oldLine[:1] + [str(newMrnaNumber)] + oldLine[2:] )\n",
      "            outputFile.write(newLine+\"\\n\")        \n",
      "        medianInitiationRate = numpy.median( initiationRateList )\n",
      "        numberOfTransgeneSequences = sum(map(lambda x: 1,Bio.SeqIO.parse( \"seq_and_annotation/Ecoli.transgene.cai\" + \"%0.1f\"%cai + \".fasta\", 'fasta' )))\n",
      "        transgeneMrnaCopyNumber = numpy.ceil( (totalNumberOfOldMrnas - totalNumberOfNewMrnas)\n",
      "                                             / float(numberOfTransgeneSequences) )\n",
      "        for record in Bio.SeqIO.parse( \"seq_and_annotation/Ecoli.transgene.cai\" + \"%0.1f\"%cai + \".fasta\", 'fasta' ):\n",
      "            outputFile.write( '%0.7f'%medianInitiationRate + ' ' +\n",
      "                                  str(int(transgeneMrnaCopyNumber)) + ' ' )\n",
      "                                  \n",
      "            for loop in range( 0, len(record)-6, 3 ):\n",
      "                outputFile.write( str( geneticCode.ix[ str(record.seq[loop:loop+3] )]['index'] ) + ' ' )\n",
      "            \n",
      "            outputFile.write( str( geneticCode.ix[ str(record.seq[-6:-3] )]['index'] ) + '\\n' ) \n",
      "        outputFile.close()        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 43
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<a id=\"source\"></a>\n",
      "### Source code for simulation ###\n",
      "\n",
      "Copy the following source code for the simulation into a file \n",
      "\n",
      "    scripts/ecoli_translation.c"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "```c\n",
      "// Source code for whole-cell simulation of E. coli protein synthesis.\n",
      "// Author: Arvind R. Subramaniam\n",
      "// Reference: Arvind R. Subramaniam and Erin O'Shea, unpublished, 2014.\n",
      "// Author version number 1.34; Date 11 May 2014.\n",
      "//\n",
      "// Source code was adapted from the yeast translation model in \n",
      "// Rate-limiting steps in yeast protein translation. \n",
      "// Shah P, Ding Y, Niemczyk M, Kudla G, and Plotkin JB. \n",
      "// Cell 153 (7): 1589-1601 (2013) \n",
      "\n",
      "///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////\n",
      "\n",
      "// Declaring Header Files\n",
      "#include <stdio.h>\n",
      "#include <stdlib.h>\n",
      "#include <math.h>\n",
      "#include <time.h>\n",
      "#include <sys/time.h> \n",
      "#include <string.h>\n",
      "\n",
      "///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////\n",
      "\n",
      "// Fixed parameters\n",
      "#define AVOGADRO_NUMBER 6.023e23\t\n",
      "#define MAX_GENE_LEN_ALLW 5000 // Required for initialization of mRNA data structure.\n",
      "\n",
      "///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////\n",
      "\n",
      "// Default global variables\n",
      "int seed = 111;\t\t\t\t\t\t\t\t// Seed for random number generator. \n",
      "double codonMinElongationTime = 0.05; \t\t// Includes peptidyl-transfer and translocation time, \n",
      "\t\t\t\t\t\t\t\t\t\t\t// but not codon-dependent tRNA selection time \n",
      "\t\t\t\t\t\t\t\t\t\t\t// which is added separately below, \n",
      "\t\t\t\t\t\t\t\t\t\t\t// command-line option provided.\n",
      "int totalNumberOfGenes = 1550;\t\t\t\t// Number of genes,\n",
      "\t\t\t\t\t\t\t\t\t\t\t// required for initiatization of data structure.\n",
      "int totalNumberOfRibosomes = 4.4e4;\t\t\t// Total ribosomes.  Command line option provided.\n",
      "double boundRibosomeFraction = 0.85;\t\t// Fraction of ribosomes bound to mRNA\n",
      "int totalNumberOfTrna = 4.08e5;\t\t\t\t// Total tRNAs\n",
      "int totalNumberOfMrna = 0;\t\t\t\t\t// Initialization,\n",
      "\t\t\t\t\t\t\t\t\t\t\t// Actual value calculated from mRNA copy number file.\n",
      "double cellVolume = 1.0e-18;\t\t\t\t// Total space within a cell\n",
      "double ribosomeTrnaKcatByKm = 2.0e8; \t\t// kcat/km for ternary complex ribosome association,\n",
      "\t\t\t\t\t\t\t\t\t\t\t// commandline option provided.\n",
      "double abortionThresholdRate = 10;\t\t\t// s-1, Abortion Rate become non-zero only if \n",
      "\t\t\t\t\t\t\t\t\t\t\t// the codon elongation rate falls below this value.\n",
      "\t\t\t\t\t\t\t\t\t\t\t// Command line option provided.\n",
      "double abortionRateConstant = 0; \t\t\t// Non-zero value implies ribosomes will falloff.\n",
      "\t\t\t\t\t\t\t\t\t\t\t// Command line option provided.\n",
      "double starvationFactor = 1; \t\t\t\t// Value below one implies that \n",
      "\t\t\t\t\t\t\t\t\t\t\t// the cell is starving for an amino acid.\n",
      "\t\t\t\t\t\t\t\t\t\t\t// Command line option provided.\n",
      "double relativeTrnaAminoacylationRates[5] \n",
      "\t\t\t\t\t\t= {1,1,1,1,1}; \t\t// These factors multiply \n",
      "\t\t\t\t\t\t\t\t\t\t\t// the aminoacylation rate constants for the tRNAs \n",
      "\t\t\t\t\t\t\t\t\t\t\t// corresponding to the liming amino acid.\n",
      "\t\t\t\t\t\t\t\t\t\t\t// Command line option provided.\n",
      "\n",
      "// Run options\n",
      "int printOptions[5] = {1,1,1,1,1}; \t\t\t// Controls what outputs are printed to files.\n",
      "\t\t\t\t\t\t\t\t\t\t\t// Look at end of main function for file details.\n",
      "char outfilePrefix[350] = \"\";\t\t\t\t// Prefix for output file names.\n",
      "\t\t\t\t\t\t\t\t\t\t\t// Common-line option provided.\n",
      "char outfile[350] = \"x\"; \t\t\t\t\t// Variable to temporarily store \n",
      "\t\t\t\t\t\t\t\t\t\t\t// the names of different output file names.\n",
      "char mrnaFile[350] = \"\"; \t\t\t\t\t// Set from command-line. \n",
      "\t\t\t\t\t\t\t\t\t\t\t// This file will contain the \n",
      "\t\t\t\t\t\t\t\t\t\t\t// initiation rate of mRNAs, \n",
      "\t\t\t\t\t\t\t\t\t\t\t// mRNA copy number in the cell, \n",
      "\t\t\t\t\t\t\t\t\t\t\t// a numerical codon sequence \n",
      "\t\t\t\t\t\t\t\t\t\t\t// for all mRNAs that are simulated.\n",
      "\n",
      "// Input Files specifying parameters for different tRNA species.\n",
      "\n",
      "char trnaConcentrationFile[150] = \"seq_and_annotation/Ecoli.trna.concentrations.tsv\"; // tRNA concentrations.\n",
      "char cognateTrnaFile[150] = \"seq_and_annotation/Ecoli.cognate.pairs.tsv\"; // Cognate tRNA for each codon.\n",
      "char pairingStrengthFile[150] = \"seq_and_annotation/Ecoli.cognate.weights.tsv\"; // Codon-tRNA interaction strength.\n",
      "char aminoacylationRateFile[150] = \"seq_and_annotation/Ecoli.trna.aminoacylation.rate.tsv\"; // Aminoacylation rates of all tRNAs.\n",
      "double totalTime = 15;\t\t\t\t\t\t// Total time for simulation.\n",
      "\t\t\t\t\t\t\t\t\t\t\t// Command line option provided.\n",
      "double thresholdTime = 10;\t\t\t\t\t// Threshold time for starting analysis.\n",
      "\t\t\t\t\t\t\t\t\t\t\t// Command line option provided.\n",
      "\n",
      "//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////\n",
      "\n",
      "// Create ribosome, mRNA, tRNA, codon structures\n",
      "typedef struct\n",
      "{\tint mRNA;\t\t\t\t\t\t\t\t// Bound to which mRNA\n",
      "\tint pos;\t\t\t\t\t\t\t\t// Position on mRNA\n",
      "\tint tRNA;\t\t\t\t\t\t\t\t// tRNA identity at the P-site\n",
      "\tdouble timeOfLastInitiation;\t\t\t// Time of translation initiation\n",
      "\tdouble timeOfLastElongation;\t\t\t// Time of arrival at current codon,\n",
      "\t\t\t\t\t\t\t\t\t\t\t// used for calculating ribosome density.\n",
      "\tdouble timeOfElongatableSwitch; \t\t// Time when the ribosome became elongatable \n",
      "\t\t\t\t\t\t\t\t\t\t\t// due to moving of a codon upfront. \n",
      "\t\t\t\t\t\t\t\t\t\t\t// Used for calculating codon elongation times.\n",
      "\tint elng_cod_list;\t\t\t\t\t\t// The id of the list of elongatable codons\n",
      "\tint elng_pos_list;\t\t\t\t\t\t// Position in the list of elongatable codons\n",
      "} ribosome;\n",
      "\n",
      "typedef struct\n",
      "{\tint gene;\t\t\t\t\t\t\t\t// Gene id\n",
      "\tint numberOfInitiations;\t\t\t\t// Number of initiation events\n",
      "\tint numberOfFullProteins;\t\t\t\t// Number of translation events\n",
      "\tint numberOfAbortedProteins;\t\t\t// Number of translation events\n",
      "\tdouble timeOfLastMrnaInitiation;\t\t// Time of last initiation event\n",
      "\tdouble averageInitiationTime;\t\t\t// Average time to initiation\n",
      "\tdouble averageElongationTime;\t\t\t// Average translation time\n",
      "\tdouble averageAbortionTime;\t\t\t\t// Average translation time\n",
      "} transcript;\n",
      "\n",
      "typedef struct\n",
      "{\tint seq[MAX_GENE_LEN_ALLW];\t\t\t\t// Codon sequence of the gene\n",
      "\tint length;\t\t\t\t\t\t\t\t// Length of the gene\n",
      "\tint exp;\t\t\t\t\t\t\t\t// Gene expression level\n",
      "\tdouble ini_prob;\t\t\t\t\t\t// Initiation probability of the mRNA\n",
      "} gene;\n",
      "\n",
      "typedef struct\n",
      "{\tint id;\t\t\t\t\t\t\t\t\t// tRNA id, ranges between 1 and 40.\n",
      "\tchar anticodon[4];\t\t\t\t\t\t// unmodified anticodon sequence\n",
      "\tchar aminoacid[4];\t\t\t\t\t\t// three letter amino acid code\n",
      "\tdouble concn;\t\t\t\t\t\t\t// tRNA concentration\n",
      "\tdouble trnaRibosomeRateConstant; \t\t// kcat/Km for the trna-ternary complex to \n",
      "\t\t\t\t\t\t\t\t\t\t\t// bind to the ribosome at the A-site.\n",
      "\tdouble aminoacylationRateConstant; \t\t\t// this can be decreased to account for charging.\n",
      "} trna;\n",
      "\n",
      "typedef struct\n",
      "{\n",
      "\tint id;\t\t\t\t\t\t\t\t\t// codon id, ranges from 0 to 60.\n",
      "\tchar triplet[4]; \t\t\t\t\t\t// codon seuquence.\n",
      "\tint trna[2];\t\t\t\t\t\t\t// cognate trna ids, range from 0 to 40. \n",
      "\t\t\t\t\t\t\t\t\t\t\t// 0 implies no second cognate trna.\n",
      "\tdouble weight[2];\t\t\t\t\t\t// codon-trna pairing weight.\n",
      "\t\t\t\t\t\t\t\t\t\t\t// 0 implies no second cognate trna.\n",
      "} codon;\n",
      "\n",
      "//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////\n",
      "\n",
      "// Reading the processed sequence file\n",
      "int Read_sequence_File(char *filename, gene *Gene)\n",
      "{   FILE *fh; \n",
      "    int c1=0,c2=0;\n",
      "    char curr_char;\n",
      "\tchar buffer[5000];\n",
      "    \n",
      "    fh=fopen(filename, \"r\");\n",
      "\n",
      "    fscanf(fh,\"%lf\",&Gene[c1].ini_prob);\n",
      "\n",
      "    do   \n",
      "    {   fscanf(fh,\"%d\",&Gene[c1].exp);\n",
      "     \n",
      "        c2 = 0; \n",
      "        do   \n",
      "        {   fscanf(fh,\"%d\",&Gene[c1].seq[c2]);\n",
      "            c2++;\n",
      "            curr_char = fgetc(fh);\n",
      "        }while(curr_char != '\\n');\n",
      "        Gene[c1].length = c2;\n",
      "        c1++;\n",
      "    }while(fscanf(fh,\"%lf\",&Gene[c1].ini_prob) ==1);\n",
      "}\n",
      "\n",
      "// Read in tRNA concentration file\n",
      "int Read_tRNA_File(char *trnaConcnFile, char *trnaAminoacylationRateFile, trna *Trna)\n",
      "{   FILE *fh1, *fh2; \n",
      "    int loop1=0;\n",
      "\tint loop2=0;\n",
      "\tchar buffer[256];\n",
      "\tint id;\n",
      "\tdouble rate;\n",
      "\t\n",
      "\tfh1=fopen(trnaConcnFile, \"r\");\n",
      "    \n",
      "\tfgets( buffer, sizeof(buffer), fh1); // Read the header line first.\n",
      "    \n",
      "\twhile(fscanf(fh1,\"%d\",&Trna[loop1].id)==1)\n",
      "\t{\t\n",
      "   \t\tfscanf(fh1,\"%s%lf%s\", &Trna[loop1].anticodon, &Trna[loop1].concn, &Trna[loop1].aminoacid);\n",
      "        loop1++;\n",
      "    }\n",
      "   \n",
      "\tbuffer[0] = 0;\n",
      " \n",
      "\tfh2=fopen(trnaAminoacylationRateFile, \"r\");\n",
      "    \n",
      "\tfgets( buffer, sizeof(buffer), fh2); // Read the header line first.\n",
      "    \n",
      "\twhile(fscanf(fh2,\"%d\",&id)==1)\n",
      "\t{\t\n",
      "\t\tchar anticodon[4]; // I do not store the anticodon. \n",
      "   \t\tfscanf(fh2,\"%s%lf\", &anticodon, &Trna[loop2].aminoacylationRateConstant);\n",
      "\t\tloop2++;\n",
      "\t}\n",
      "\n",
      "}\n",
      "\n",
      "// Read in codon-trna pairs and codon-trna weights. \n",
      "int Read_Codon_Files(char *cognateTrnaFile, char *pairingStrengthFile, codon *Codon) \n",
      "{   FILE *pair_file;\n",
      "\tFILE *weight_file; \n",
      "    int loop1=0;\n",
      "\tchar buffer[256];\n",
      "    \n",
      "    pair_file=fopen(cognateTrnaFile, \"r\");\n",
      "    \n",
      "    if(!pair_file)                 // Check if file exists\n",
      "    {   printf(\"\\nPair file doesn't exist.\\n\");\n",
      "        exit(1);\n",
      "    }    \n",
      "\t\n",
      "    weight_file=fopen(pairingStrengthFile, \"r\");\n",
      "    \n",
      "    if(!weight_file)                 // Check if file exists\n",
      "    {   printf(\"\\nWeight file doesn't exist.\\n\");\n",
      "        exit(1);\n",
      "    }    \n",
      "\t\n",
      "\tfgets( buffer, sizeof(buffer), pair_file); // Read the header line first.\n",
      "\tfgets( buffer, sizeof(buffer), weight_file); // Read the header line first.\n",
      "    \n",
      "\twhile(fscanf(pair_file,\"%d\",&Codon[loop1].id)==1)\n",
      "\t{\t\n",
      "\t\tchar codon1[4], codon2[4]; // I do not store the codon. \n",
      "   \t\tint temp_id = 0;\n",
      "\t\tfscanf(pair_file,\"%s%d%d\", &Codon[loop1].triplet, &Codon[loop1].trna[0], &Codon[loop1].trna[1]);\n",
      "   \t\tfscanf(weight_file,\"%d%s%lf%lf\", &temp_id, &codon2, &Codon[loop1].weight[0], &Codon[loop1].weight[1]);\n",
      "        loop1++;\n",
      "    }    \n",
      "}\n",
      "\n",
      "\n",
      "// Read in commandline arguments\n",
      "void Read_Commandline_Args(int argc, char *argv[], trna *Trna)\n",
      "\t{\t\n",
      "\tint argIndex;\n",
      "\tFILE *fh;\n",
      "\tint c1;\n",
      "\tint trnaloop;\n",
      "\tfor(argIndex=1;argIndex<argc;argIndex++)\n",
      "\t\t{\n",
      "\t\tif ( !strcmp(argv[argIndex], \"--output-prefix\") )\n",
      "\t\t\t{\n",
      "\t\t\tstrcat( outfilePrefix, argv[argIndex+1] );\t\n",
      "\t\t\t}\n",
      "\t\tif ( !strcmp(argv[argIndex], \"--input-genes\") )\n",
      "\t\t\t{\n",
      "\t\t\tstrcat( mrnaFile, argv[argIndex+1] );\t\n",
      "\t\t\t}\n",
      "\t\t}\n",
      "\tfor(argIndex=1;argIndex<argc;argIndex++)\n",
      "\t\t{\n",
      "\t\tif ( !strcmp(argv[argIndex], \"--total-ribosomes\") )\n",
      "\t\t\t{\n",
      "\t\t\ttotalNumberOfRibosomes = atoi( argv[argIndex+1] );\n",
      "\t\t\tstrcat( outfilePrefix, \"_ribosomes_\" );\n",
      "\t\t\tstrcat( outfilePrefix, argv[argIndex+1] );\t\n",
      "\t\t\t}\n",
      "\t\tif ( !strcmp(argv[argIndex], \"--ribotrnakcatkm\") )\n",
      "\t\t\t{\n",
      "\t\t\tribosomeTrnaKcatByKm = atof( argv[argIndex+1] );\n",
      "\t\t\tstrcat( outfilePrefix, \"_ribotrnakcatkm_\" );\n",
      "\t\t\tstrcat( outfilePrefix, argv[argIndex+1] );\n",
      "\t\t\t}\n",
      "\t\tif ( !strcmp(argv[argIndex], \"--abort-threshold-rate\") )\n",
      "\t\t\t{\n",
      "\t\t\tabortionThresholdRate = atof( argv[argIndex+1] );\n",
      "\t\t\tstrcat( outfilePrefix, \"_abortthreshold_\" );\n",
      "\t\t\tstrcat( outfilePrefix, argv[argIndex+1] );\n",
      "\t\t\t}\n",
      "\t\tif ( !strcmp(argv[argIndex], \"--min-elong-time\") )\n",
      "\t\t\t{\n",
      "\t\t\tcodonMinElongationTime = atof( argv[argIndex+1] );\n",
      "\t\t\tstrcat( outfilePrefix, \"_minelongtime_\" );\n",
      "\t\t\tstrcat( outfilePrefix, argv[argIndex+1] );\n",
      "\t\t\t}\n",
      "\t\tif ( !strcmp(argv[argIndex], \"--aminoacid\") )\n",
      "\t\t\t{\n",
      "\t\t\tstrcat( outfilePrefix, \"_\" );\n",
      "\t\t\tstrcat( outfilePrefix, argv[argIndex+1] ); \n",
      "\t\t\tif ( !strcmp(argv[argIndex+2], \"starvationfactor\" ) )\n",
      "\t\t\t\t{\n",
      "\t\t\t\tstarvationFactor = atof(argv[argIndex+3]);\n",
      "\t\t\t\tstrcat( outfilePrefix, \"_starvationfactor_\" );\n",
      "\t\t\t\tstrcat( outfilePrefix, argv[argIndex+3] ); \n",
      "\t\t\t\tfor(trnaloop=0;trnaloop<61;trnaloop++)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tif ( !strcmp( Trna[trnaloop].aminoacid, argv[argIndex+1] ) )\n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tTrna[trnaloop].aminoacylationRateConstant *= starvationFactor;\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t\t\t\t\n",
      "\t\t\tif ( !strcmp(argv[argIndex+2], \"totalcognatetrnafactor\" ) )\n",
      "\t\t\t\t{\n",
      "\t\t\t\tstarvationFactor = atof(argv[argIndex+3]);\n",
      "\t\t\t\tstrcat( outfilePrefix, \"_totalcognatetrnafactor_\" );\n",
      "\t\t\t\tstrcat( outfilePrefix, argv[argIndex+3] );\n",
      "\t\t\t\tfor(trnaloop=0;trnaloop<61;trnaloop++)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tif (!strcmp( Trna[trnaloop].aminoacid, argv[argIndex+1] ))\n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tTrna[trnaloop].concn *= starvationFactor;\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\n",
      "\t\t\tif ( !strcmp(argv[argIndex+2], \"relativetrnaaminoacylationrates\" ) )\n",
      "\t\t\t\t{\n",
      "\t\t\t\tstrcat( outfilePrefix, \"_relativetrnaaminoacylationrates\" );\n",
      "\t\t\t\tc1 = 0;\n",
      "\t\t\t\twhile (argv[c1+argIndex+3][0] != '-' )\n",
      "\t\t\t\t\t{\n",
      "                    if (c1+argIndex+3 == argc-1)\n",
      "                        {\n",
      "                        break;\n",
      "                        }\n",
      "\t\t\t\t\trelativeTrnaAminoacylationRates[c1] = atof(argv[c1+argIndex+3]);\n",
      "\t\t\t\t\tstrcat( outfilePrefix, \"_\" );\n",
      "\t\t\t\t\tstrcat( outfilePrefix, argv[c1+argIndex+3] );\n",
      "\t\t\t\t\tc1 += 1;\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\tc1 = 0;\n",
      "\t\t\t\tfor(trnaloop=0;trnaloop<61;trnaloop++)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tif ( !strcmp( Trna[trnaloop].aminoacid, argv[argIndex+1] ) )\n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tTrna[trnaloop].aminoacylationRateConstant *= relativeTrnaAminoacylationRates[c1];\n",
      "\t\t\t\t\t\tc1 += 1;\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t}\n",
      "\n",
      "\t\tif ( !strcmp(argv[argIndex], \"--abortion-relative-rate\") )\n",
      "\t\t\t{\t\n",
      "\t\t\tabortionRateConstant = atof(argv[argIndex+1]);\t\n",
      "\t\t\tstrcat( outfilePrefix, \"_abortion_\" );\n",
      "\t\t\tstrcat( outfilePrefix, argv[argIndex+1] );\t\n",
      "\t\t\t}\n",
      "\n",
      "\t\tif ( !strcmp(argv[argIndex], \"--threshold-time\") )\n",
      "\t\t\t{\t\n",
      "\t\t\tthresholdTime = atof(argv[argIndex+1]);\t\n",
      "\t\t\tstrcat( outfilePrefix, \"_thresholdtime_\" );\n",
      "\t\t\tstrcat( outfilePrefix, argv[argIndex+1] );\t\n",
      "\t\t\t}\n",
      "\n",
      "\t\tif ( !strcmp(argv[argIndex], \"--total-time\") )\n",
      "\t\t\t{\t\n",
      "\t\t\ttotalTime = atof(argv[argIndex+1]);\t\n",
      "\t\t\tstrcat( outfilePrefix, \"_totaltime_\" );\n",
      "\t\t\tstrcat( outfilePrefix, argv[argIndex+1] );\t\n",
      "\t\t\t}\n",
      "\t\t}\n",
      "\n",
      "\tstrcpy( outfile, outfilePrefix);\n",
      "\tfh = fopen( strcat( outfile, \"_args.in\" ), \"w\" );\n",
      "\tfprintf(fh, \"%s \", argv[0] );\n",
      "\tfor(argIndex=1;argIndex<argc;argIndex++)\n",
      "\t\t{\n",
      "\t\tfprintf(fh, \"%s \", argv[argIndex] );\n",
      "\t\t}\n",
      "\tfclose(fh);\n",
      "}\n",
      "\n",
      "\n",
      "//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////\n",
      "\n",
      "// Main Function\n",
      "int main(int argc, char *argv[])\n",
      "{\t\n",
      "\tint c1, c2, c3, c4;\n",
      "\tstruct timeval time1;\n",
      "\tstruct timeval time2;\n",
      "\tFILE *f1, *f2, *f3, *f4, *f5, *f6, *f6a, *f7, *f8;\n",
      "\n",
      "\tgettimeofday(&time1, NULL);\n",
      "\n",
      "\tsrand(seed);\n",
      "\t\t\n",
      "\t// Initialize the various structures\n",
      "\tgene *Gene = (gene *)malloc(sizeof(gene) * totalNumberOfGenes);\n",
      "\tif(Gene == NULL)\n",
      "\t{\tprintf(\"Too many genes\\nOut of memory\\n\");fflush(stdout);\n",
      "\t\texit(1);\n",
      "\t}\n",
      "\t\n",
      "\tribosome *Ribosome = (ribosome *)malloc(sizeof(ribosome) * totalNumberOfRibosomes);\n",
      "\tif(Ribosome == NULL)\n",
      "\t{\tprintf(\"Too many ribosomes\\nOut of memory\\n\");fflush(stdout);\n",
      "\t\texit(1);\n",
      "\t}\n",
      "\t\t\n",
      "\ttrna *Trna = (trna *)malloc(sizeof(trna) * 61);\n",
      "\tif(Trna == NULL)\n",
      "\t{\tprintf(\"Too many tRNAs\\nOut of memory\\n\");fflush(stdout);\n",
      "\t\texit(1);\n",
      "\t}\n",
      "\n",
      "\tcodon *Codon = (codon *)malloc(sizeof(codon) * 61);\n",
      "\tif(Codon == NULL)\n",
      "\t{\tprintf(\"Too many codons\\nOut of memory\\n\");fflush(stdout);\n",
      "\t\texit(1);\n",
      "\t}\n",
      "\n",
      "//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////\n",
      "\t\n",
      "\t\n",
      "\t// Read in the trna concentrations and aminoacylation rates;\n",
      "\tRead_tRNA_File(trnaConcentrationFile, aminoacylationRateFile, Trna);\n",
      "\n",
      "\t// Read in the codon trna pairings and their corresponding weights;\n",
      "\tRead_Codon_Files(cognateTrnaFile, pairingStrengthFile, Codon);\n",
      "\n",
      "\tRead_Commandline_Args( argc, argv, Trna );\n",
      "\n",
      "\t// Read in the numeric seq\n",
      "\tRead_sequence_File(mrnaFile, Gene);\n",
      "\n",
      "\tprintf(\"Read input files succcessfully.\\nTotal run time = %0.0fs.\\n////////////////////////\\n\",totalTime);\n",
      "\n",
      "\t// Calculate the total number of mRNAs in the cell based on the second col in the fasta file.\n",
      "    totalNumberOfMrna = 0;\n",
      "\tint geneloop;\n",
      "    for(geneloop=0;geneloop<totalNumberOfGenes;geneloop++)\n",
      "    {   totalNumberOfMrna += Gene[geneloop].exp;\n",
      "    }   \n",
      "    \n",
      "    transcript *mRNA = (transcript *)malloc(sizeof(transcript) * totalNumberOfMrna);\n",
      "    if(mRNA == NULL)\n",
      "    {   printf(\"Too many mRNAs\\nOut of memory\\n\");fflush(stdout);\n",
      "        exit(1);\n",
      "    }   \n",
      "\n",
      "    int mrnaloop=0;\n",
      "\tint exprloop;\n",
      "    for(geneloop=0;geneloop<totalNumberOfGenes;geneloop++)              // Fill the mRNA struct with\n",
      "    {   for(exprloop=0;exprloop<Gene[geneloop].exp;exprloop++)          // length and ini_rate info.\n",
      "        {   mRNA[mrnaloop].gene = geneloop; \n",
      "            mRNA[mrnaloop].numberOfInitiations = 0;\n",
      "            mRNA[mrnaloop].timeOfLastMrnaInitiation = 0.0;\n",
      "            mRNA[mrnaloop].numberOfFullProteins = 0;\n",
      "            mRNA[mrnaloop].numberOfAbortedProteins = 0;\n",
      "            mRNA[mrnaloop].averageInitiationTime = 0.0;\n",
      "            mRNA[mrnaloop].averageElongationTime = 0.0;\n",
      "            mRNA[mrnaloop].averageAbortionTime = 0.0;\n",
      "            mrnaloop++;\n",
      "        }   \n",
      "    }  \t\n",
      "\n",
      "    // Variables to track the translation process\n",
      "    int deacylatedFreeTrna[61];\t\t\t\t\t// Number of uncharged free trna of each type.\n",
      "\tint acylatedFreeTrna[61];                   // Number of chaged free tRNAs of each type.\n",
      "    int ribosomeBoundTrna[61];                  // Number of ribosome bound tRNAs of each type.\n",
      "\tint totalTrna[61];\t\t\t\t\t\t\t// Total number of tRNAs of type.\n",
      "    int Rf = totalNumberOfRibosomes;            // Total number of free ribosomes\n",
      "\t\t\t\t\t\t\t\t\t\t\t\t// (at start all ribosomes are free)\n",
      "    int Rb[61];                                 // Number of ribosomes bound to each codon\n",
      "    int freeMrna[totalNumberOfGenes];                 // Number of initiable mRNAs of each gene\n",
      "\t\t\t\t\t\t\t\t\t\t\t\t// (first 10 codons unbound by ribosomes)\n",
      "    int totalFreeMrna;                                 // Total number of initiable mRNAs\n",
      "    int x;\n",
      "    int n_Rb_e[61];                             // Number of elongable bound ribosomes to each codon\n",
      "    int r_id; \t\t\t\t\t\t\t\t\t// Current ribosome id.\n",
      "    int m_id;\t\t\t\t\t\t\t\t\t// Current mRNA id.\n",
      "    int c_id;\t\t\t\t\t\t\t\t\t// Current codon id.\n",
      "\tint cognate_trna_id = 100;\t\t\t\t\t// Current cognate tRNA id.\n",
      "    int c2_id;\n",
      "    int g_id;\n",
      "\tdouble coin, prob_g;\n",
      "    int obs_max_len = 0;                        // Observed max gene length\n",
      "    int obs_max_exp = 0;                        // Observed max gene expression\n",
      "    int next_avail_ribo = 0;\n",
      "    int termtn_now=0;\n",
      "    int numberOfElongations[61];                // Number of times a codon type is elongated\n",
      "    int numberOfAbortions[61];                  // Number of times a codon type is elongated\n",
      "    double summedElongationTimes[61];\n",
      "    double initiationScaledMrnaCopyNumber[totalNumberOfGenes];\n",
      "    double totalInitiationScaledMrnaCopyNumber = 0.0;\n",
      "    double totalInitiationRate, \n",
      "\t\t   totalElongationRate, \n",
      "\t\t   totalAminoacylationRate;\n",
      "\tdouble r_elng[61];\n",
      "\tdouble avgAcylatedTrnaConcn[61], \n",
      "\t\t   avgDeacylatedTrnaConcn[61], \n",
      "\t\t   avgRibosomeBoundTrnaConcn[61]; \t\t// Average number of free tRNAs of each type\n",
      "\t\t\t\t\t\t\t\t\t\t\t\t// (averaged by time)\n",
      "\tdouble avg_Rf = {0.0};\t\t\t\t\t\t// Average number of free ribosomes\n",
      "\t\t\t\t\t\t\t\t\t\t\t\t// (averaged by time)\n",
      "\n",
      "//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////\n",
      "\n",
      "// Begin simulation of the translation process\n",
      "\tint abort_flag; \t\t\t\t\t\t\t// Will be set to 1 if the ribosome \n",
      "\t\t\t\t\t\t\t\t\t\t\t\t// is aborting at a particular codon.\n",
      "\tdouble currentTime = 0.0;\n",
      "\tdouble previousRate, currentRate;\n",
      "\tdouble totalRate = 0.0;\n",
      "\tdouble randomRate;\n",
      "\tdouble randomAbortionRate = 0.0;\n",
      "\tint t_print = floor(thresholdTime);\n",
      "\tchar tmp_t[10];\n",
      "\tint number_of_iterations = 0;\t\n",
      "\tdouble codon_dependent_elongation_rate[61];\n",
      "\tdouble abortionRate[61];\n",
      "\tdouble codon_elongation_rate[61];\n",
      "\tint cognatetrnaloop;\n",
      "\tint temp_trna_total, temp_trna;\t\n",
      "\tint **Rb_e;\t\t\t\t\t\t\t\t\t// Bound ribosomes to each codon that can be elongated\n",
      "\tRb_e = malloc(sizeof(int *) * 62);\n",
      "\n",
      "\tfor(c1=0;c1<62;c1++)\n",
      "\t{\tRb_e[c1] = malloc(sizeof(int *) * totalNumberOfRibosomes);\n",
      "\t}\n",
      "\t\n",
      "\tfor(c1=0;c1<totalNumberOfGenes;c1++)\n",
      "\t{\tif(Gene[c1].exp > obs_max_exp)\n",
      "\t\t{\tobs_max_exp = Gene[c1].exp;\n",
      "\t\t}\n",
      "\t\tif(Gene[c1].length > obs_max_len)\n",
      "\t\t{\tobs_max_len = Gene[c1].length;\n",
      "\t\t}\n",
      "\t\tfreeMrna[c1] = Gene[c1].exp;\n",
      "\t\ttotalFreeMrna += freeMrna[c1];\n",
      "\t\tinitiationScaledMrnaCopyNumber[c1] = freeMrna[c1]*Gene[c1].ini_prob;\t// Scale gene_exp with ini_prob\n",
      "\t\ttotalInitiationScaledMrnaCopyNumber += initiationScaledMrnaCopyNumber[c1];\n",
      "\t\t\n",
      "\t}\n",
      "\n",
      "\tint **freeMrnaList;\t\t\t\t\t\t\t// List to figure out which mRNAs can be initiated\n",
      "\t\t\t\t\t\t\t\t\t\t\t\t// based on no bound ribosomes from pos=0->pos=10\n",
      "\tfreeMrnaList = malloc(sizeof(int *) * totalNumberOfGenes);\n",
      "\tfor(c1=0;c1<totalNumberOfGenes;c1++)\n",
      "\t{\t\n",
      "\t\tfreeMrnaList[c1] = malloc(sizeof(int *) * obs_max_exp);\n",
      "\t}\n",
      "\n",
      "\t// Initialize R_grid and avgRibosomeDensityGrid. \n",
      "\t// R_grid now contains the id of ribosome at each mRNA position.\n",
      "\t// If there is no ribosome then that position get the value totalNumberOfRibosomes instead of 0 as 0 is a ribosome id\n",
      "\t// The state of the system with respect to mRNAs and bound ribosomes\n",
      "\t// avgRibosomeDensityGrid keeps track of the average ribosome density at each position of each mRNA after threshold time has elapsed.\n",
      "\t\n",
      "\tint** R_grid;\t\t\t\t\t\t\t\n",
      "\tint** numberOfElongationsGrid;\t\t\t\t\t\t\t\n",
      "\tfloat** avgRibosomeDensityGrid;\t\t\t\t\t\t\t\n",
      "\tR_grid = malloc(sizeof(int *) * totalNumberOfMrna);\n",
      "\tnumberOfElongationsGrid = malloc(sizeof(int *) * totalNumberOfMrna);\n",
      "\tavgRibosomeDensityGrid = (float**) malloc(sizeof(float *) * totalNumberOfMrna);\n",
      "\tfor(c1=0;c1<totalNumberOfMrna;c1++)\n",
      "\t\t{\t\n",
      "\t\tR_grid[c1] = malloc(sizeof(int *) * obs_max_len);\n",
      "\t\tnumberOfElongationsGrid[c1] = malloc(sizeof(int *) * obs_max_len);\n",
      "\t\tavgRibosomeDensityGrid[c1] = (float*) malloc(sizeof(float *) * obs_max_len);\n",
      "\t\tfor(c2=0;c2<Gene[mRNA[c1].gene].length;c2++)\n",
      "\t\t\t{\t\n",
      "\t\t\tR_grid[c1][c2] = totalNumberOfRibosomes;\n",
      "\t\t\tavgRibosomeDensityGrid[c1][c2] = 0.0;\n",
      "\t\t\tnumberOfElongationsGrid[c1][c2] = 0;\n",
      "\t\t\t}\n",
      "\t\t}\n",
      "\t\n",
      "\t// Initialize free mRNAs\n",
      "\tc3=0;\n",
      "\tfor(c2=0;c2<totalNumberOfGenes;c2++)\n",
      "\t{\tfor(c1=0;c1<Gene[c2].exp;c1++)\n",
      "\t\t{\tfreeMrnaList[c2][c1]=c3;\t\t\t\t\t\t\t\t\t\t\t// At the begininning all mRNAs can be initiated (store their ids here)\n",
      "\t\t\tc3++;\n",
      "\t\t}\n",
      "\t}\n",
      "\t\n",
      "\t// Initialize elongation time vectors and number of elongatable bound ribosomes to each codon.\n",
      "\tint codonloop;\n",
      "\tfor(codonloop=0;codonloop<61;codonloop++)\n",
      "\t{\t\n",
      "\t\tsummedElongationTimes[codonloop] = 0.0;\n",
      "\t\tnumberOfElongations[codonloop] = 0;\n",
      "\t\tnumberOfAbortions[codonloop] = 0;\n",
      "\t\tn_Rb_e[codonloop] = 0;\t\t\t\t\t\t\t\t\t\t\t\t\t\t\n",
      "\t}\n",
      "\n",
      "\t// Scale tRNA concentration to match the total number of tRNAs at that growth rate.\n",
      "\t// Scale tRNA aminoacylation rate and ribosome associate rate to change units.\n",
      "\tint trnaloop;\n",
      "\tdouble temp_trnatotal = 0;\n",
      "\tfor(trnaloop=0;trnaloop<61;trnaloop++)\n",
      "\t{\t\n",
      "\t\ttemp_trnatotal += Trna[trnaloop].concn;\n",
      "\t}\n",
      "\tfor(trnaloop=0;trnaloop<61;trnaloop++)\n",
      "\t{\t\n",
      "\t\tTrna[trnaloop].concn *= (totalNumberOfTrna / temp_trnatotal);\n",
      "\t\tTrna[trnaloop].aminoacylationRateConstant = Trna[trnaloop].aminoacylationRateConstant / (cellVolume * 1000 * AVOGADRO_NUMBER); // converting from (moles per litre)-1s-1 to (molecules per cell)-1s-1 units.\n",
      "\t\tTrna[trnaloop].trnaRibosomeRateConstant = ribosomeTrnaKcatByKm / (cellVolume * 1000 * AVOGADRO_NUMBER); // converting from (moles per litre)-1s-1 to (molecules per cell)-1s-1 units.\n",
      "\t}\n",
      "\n",
      "\t// Initialize the pools of deacylated, acylated and ribosome bound tRNAs.\n",
      "\t// All trnas are initially free and deacylated.\n",
      "\tfor(trnaloop=0;trnaloop<61;trnaloop++)\n",
      "\t{\t\n",
      "\t\tdeacylatedFreeTrna[trnaloop] = Trna[trnaloop].concn;\n",
      "\t\tacylatedFreeTrna[trnaloop] = 0;\n",
      "\t\tribosomeBoundTrna[trnaloop] = 0;\n",
      "\t\ttotalTrna[trnaloop] = deacylatedFreeTrna[trnaloop] + acylatedFreeTrna[trnaloop] + ribosomeBoundTrna[trnaloop];\n",
      "\t\tavgAcylatedTrnaConcn[trnaloop] = 0.0;\n",
      "\t\tavgDeacylatedTrnaConcn[trnaloop] = 0.0;\n",
      "\t\tavgRibosomeBoundTrnaConcn[trnaloop] = 0.0;\n",
      "\t}\n",
      "\n",
      "\t/////////////////////////////////////////////////\n",
      "\t// Begin the actual continuous time simulation\n",
      "\t/////////////////////////////////////////////////\n",
      "\n",
      "\twhile(currentTime<totalTime)\t\t\t\t\t\t\t\t\t\t\t\t\t\t// Till current time is less than max time\n",
      "\t{\n",
      "\t\t// Calculate the total rate of Gillespie simulation at this step as the sum of the rates of initiation, elongation, aminoacylation and abortion.\n",
      "\t\ttotalRate = 0.0;\n",
      "\n",
      "\t\t// Total Aminoacylation Rate\n",
      "\t\ttotalAminoacylationRate = 0.0;\n",
      "\t\tfor(trnaloop=0;trnaloop<61;trnaloop++)\n",
      "\t\t\t{\t\n",
      "\t\t\ttotalAminoacylationRate += ( Trna[trnaloop].aminoacylationRateConstant * deacylatedFreeTrna[trnaloop] );\n",
      "\t\t\t}\n",
      "\t\ttotalRate += totalAminoacylationRate;\n",
      "\n",
      "\t\t// Total Initiation Rate\n",
      "\t\ttotalInitiationRate = totalInitiationScaledMrnaCopyNumber*Rf/( (1-boundRibosomeFraction) * totalNumberOfRibosomes );\t\t\t\t\n",
      "\t\ttotalRate += totalInitiationRate;\n",
      "\n",
      "\t\t// Total Elongation and Abortion Rates\n",
      "\t\ttotalElongationRate = 0.0;\n",
      "\t\tfor(codonloop=0;codonloop<61;codonloop++)\n",
      "\t\t{\t\n",
      "\t\t\tcodon_dependent_elongation_rate[codonloop] = 0.0;\n",
      "\t\t\tcodon_elongation_rate[codonloop] = 0.0;\n",
      "\t\t\tabortionRate[codonloop] = 0.0;\n",
      "\t\t\tfor (cognatetrnaloop=0; cognatetrnaloop<2; cognatetrnaloop++)\n",
      "\t\t\t\t{\n",
      "\t\t\t\tcodon_dependent_elongation_rate[codonloop] += ( Trna[ Codon[codonloop].trna[cognatetrnaloop] ].trnaRibosomeRateConstant * acylatedFreeTrna[ Codon[codonloop].trna[cognatetrnaloop] ] * Codon[codonloop].weight[cognatetrnaloop] );\n",
      "\t\t\t\t}\n",
      "\t\t\tcodon_elongation_rate[codonloop] = 1 / (codonMinElongationTime + 1 / codon_dependent_elongation_rate[codonloop] );\n",
      "\n",
      "\t\t\tif ( codon_elongation_rate[codonloop] < abortionThresholdRate  ) // Abortion rate is non-zero only if the codon_elongation_rate falls below the threshold for abortion. \n",
      "\t\t\t\t{\n",
      "\t\t\t\tabortionRate[codonloop] = abortionRateConstant; \n",
      "\t\t\t\t}\n",
      "\n",
      "\t\t\tr_elng[codonloop] = ( codon_elongation_rate[codonloop] + abortionRate[codonloop] ) * n_Rb_e[codonloop];\n",
      "\t\t\ttotalElongationRate += r_elng[codonloop];\n",
      "\t\n",
      "\t\t}\n",
      "\t\ttotalRate += totalElongationRate;\n",
      "\n",
      "\n",
      "\n",
      "\t\t// The totalTrna values below should change at all. This condition can be useful for debugging.\n",
      "\t\ttemp_trna_total = 0;\t\n",
      "\t\tfor(trnaloop=0;trnaloop<61;trnaloop++)\n",
      "\t\t{\t\n",
      "\t\t\ttotalTrna[trnaloop] = deacylatedFreeTrna[trnaloop] + acylatedFreeTrna[trnaloop] + ribosomeBoundTrna[trnaloop];\n",
      "\t\t\ttemp_trna_total += totalTrna[trnaloop];\n",
      "\t\t}\n",
      "\t\ttemp_trna = temp_trna_total;\n",
      "\n",
      "\t\t// Increment time. Following Shah et al, we have not implemented the strict Gillespie prescription of random time drawn from an exponential distribution. Since all calculated quantities are steady-state values, this difference does not matter.\n",
      "\t\t//\n",
      "\t\tcurrentTime +=  ( 1 / totalRate );\n",
      "\n",
      "\t\t// Tracking time-averaged free ribosomes and tRNAs for final output.\n",
      "\t\tif(currentTime>thresholdTime)\n",
      "\t\t{\tif(printOptions[3]==1)\t\t\t\t\n",
      "\t\t\t\t{\t\n",
      "\t\t\t\tfor(trnaloop=0;trnaloop<61;trnaloop++)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tavgAcylatedTrnaConcn[trnaloop] += (double)acylatedFreeTrna[trnaloop] / totalRate;\n",
      "\t\t\t\t\tavgDeacylatedTrnaConcn[trnaloop] += (double)deacylatedFreeTrna[trnaloop] / totalRate;\n",
      "\t\t\t\t\tavgRibosomeBoundTrnaConcn[trnaloop] += (double)ribosomeBoundTrna[trnaloop] / totalRate;\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\tavg_Rf += (double)Rf / totalRate;\n",
      "\t\t\t\t}\n",
      "\t\t}\n",
      "\t\t\n",
      "\t\t// Draw a random reaction channel by first choosing between aminoacylation, initiation or elongation and then picking the sub channel according to the respective rate.\n",
      "\t\t\t\n",
      "\t\trandomRate = totalRate * rand()/((double)RAND_MAX);\t// Pick a random uniform to pick an event\n",
      "\t\tpreviousRate = 0.0;\n",
      "\t\tcurrentRate = 0.0;\n",
      "\n",
      "\t\t// Aminoacylation\n",
      "\t\tpreviousRate = currentRate;\n",
      "\t\tcurrentRate += totalAminoacylationRate;\t\t\t\t\t\t\t\n",
      "\t\t// Pick tRNA that is aminoacylated with a probability proportional to its aminoacylation rate.\t\n",
      "\t\tif(randomRate > previousRate && randomRate < currentRate)\n",
      "\t\t\t{\n",
      "\t\t\tfor(trnaloop=0;trnaloop<61;trnaloop++)\t\n",
      "\t\t\t\t{\t\n",
      "\t\t\t\tpreviousRate += ( Trna[trnaloop].aminoacylationRateConstant * deacylatedFreeTrna[trnaloop] );\n",
      "\t\t\t\tif(randomRate < previousRate)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tdeacylatedFreeTrna[trnaloop]--; // Decrease deacylated trna by one.\n",
      "\t\t\t\t\tacylatedFreeTrna[trnaloop]++; // Increase acylated trna by one.\n",
      "\t\t\t\t\tbreak;\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t}\n",
      "\n",
      "\t\t// Translation initiation\n",
      "\t\tpreviousRate = currentRate;\n",
      "\t\tcurrentRate += totalInitiationRate;\t\t\t\t\t\t\t\n",
      "\t\n",
      "\t\tif(randomRate > previousRate && randomRate < currentRate)\n",
      "\t\t{\t\n",
      "\t\t\t// Ribosomes are picked sequentially\n",
      "\t\t\tr_id = next_avail_ribo;\n",
      "\t\t\tRibosome[r_id].pos = 0;\n",
      "\t\t\tRibosome[r_id].timeOfLastInitiation = currentTime;\n",
      "\t\t\tRibosome[r_id].timeOfLastElongation = currentTime;\n",
      "\t\t\tRibosome[r_id].timeOfElongatableSwitch = currentTime;\n",
      "\t\t\tnext_avail_ribo++;\n",
      "\t\t\t\n",
      "\t\t\t// Pick the gene for mRNA based on their initiation rates.\n",
      "\t\t\tfor(geneloop=0;geneloop<totalNumberOfGenes;geneloop++)\n",
      "\t\t\t\t{\t\n",
      "\t\t\t\tpreviousRate += initiationScaledMrnaCopyNumber[geneloop] *Rf / ( (1-boundRibosomeFraction) * totalNumberOfRibosomes );\t\t\t\t\n",
      "\n",
      "\t\t\t\tif(randomRate < previousRate)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tc2 = rand()%freeMrna[geneloop];\t\t\t\t\t\t\t\t\t\t// Once a gene is selected pick a random mRNA\n",
      "\t\t\t\t\tm_id = freeMrnaList[geneloop][c2];\n",
      "\t\t\t\t\tfreeMrna[geneloop]--;\n",
      "\t\t\t\t\ttotalFreeMrna--;\n",
      "\t\t\t\t\tinitiationScaledMrnaCopyNumber[geneloop]-=Gene[geneloop].ini_prob;\n",
      "\t\t\t\t\ttotalInitiationScaledMrnaCopyNumber-=Gene[geneloop].ini_prob;\n",
      "\n",
      "\t\t\t\t\tif(c2!=freeMrna[geneloop])\n",
      "\t\t\t\t\t\t{\t\n",
      "\t\t\t\t\t\tfreeMrnaList[geneloop][c2] = freeMrnaList[geneloop][freeMrna[geneloop]];\t\t\t// Accounting of free mRNAs for next round\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\tbreak;\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\t\t\t\n",
      "\t\t\t\n",
      "\t\t\tif(currentTime>thresholdTime)\n",
      "\t\t\t{\tmRNA[m_id].numberOfInitiations++;\t\t\t\t\t\t\t\t\t\t\t// Store the # of initn events on this mRNA\n",
      "\t\t\t\tmRNA[m_id].averageInitiationTime += currentTime-mRNA[m_id].timeOfLastMrnaInitiation;\t\t// Time between initn\n",
      "\t\t\t}\t\n",
      "\t\t\tmRNA[m_id].timeOfLastMrnaInitiation = currentTime;\n",
      "\n",
      "\t\t\tc_id = Gene[mRNA[m_id].gene].seq[0];\t\t\t\t\t\t\t// Codon at pos 1\n",
      "\t\n",
      "\t\t\tRibosome[r_id].mRNA = m_id;\t\t\t\t\t\t\t\n",
      "\n",
      "\t\t\tif(R_grid[m_id][10]==totalNumberOfRibosomes)\t\t\t\t\t\t\t\t\t// Check if the current ribosome can be elongated\n",
      "\t\t\t{\tRb_e[c_id][n_Rb_e[c_id]] = r_id;\t\t\t\t\t\t\t// If no ribosome at pos+10 then it can\n",
      "\t\t\t\tRibosome[r_id].elng_cod_list = c_id;\n",
      "\t\t\t\tRibosome[r_id].elng_pos_list = n_Rb_e[c_id];\n",
      "\t\t\t\tn_Rb_e[c_id]++;\n",
      "\t\t\t}\n",
      "\t\t\t\n",
      "\t\t\tR_grid[m_id][0] = r_id;\t\t\t\t\t\t\t\t\t\t\t\t// Update the ribosome grid upon initiation\n",
      "\t\t\t\n",
      "\t\t\tRf--;\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t// Update number of free ribosomes\n",
      "\t\t}\n",
      "\t\t\n",
      "\t\t// Translation Elongation\n",
      "\t\tpreviousRate = currentRate;\n",
      "\t\tcurrentRate += totalElongationRate;\t\t\t\t\t\t\t\n",
      "\t\n",
      "\t\tif(randomRate > previousRate && randomRate < currentRate)\n",
      "\t\t{\n",
      "\t\t\t// Pick a random codon based on respective ( elongation + abortion rate )\t\n",
      "\t\t\tfor(codonloop=0;codonloop<61;codonloop++)\t\n",
      "\t\t\t\t{\t\n",
      "\t\t\t\tpreviousRate += r_elng[codonloop];\n",
      "\t\t\t\tif(randomRate < previousRate)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tc_id = codonloop;\n",
      "\t\t\t\t\tbreak;\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t\n",
      "\t\t\t// Pick between elongation and abortion based on the respective probability if abortion rate is non-zero.\n",
      "\t\t\tabort_flag = 0;\n",
      "\t\t\tif (abortionRate[c_id] > 0)\n",
      "\t\t\t\t{\n",
      "\t\t\t\trandomAbortionRate = (codon_elongation_rate[c_id] + abortionRate[c_id] ) * rand()/((double)RAND_MAX); \n",
      "\t\t\t\tif ( randomAbortionRate < abortionRate[c_id] )\n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\tabort_flag = 1;\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "    \n",
      "\t\t\tx = rand()%n_Rb_e[c_id];\t\t\t\t\t\t\t\t\t\t\t// Randomly pick an elongatable ribosome bound to codon c_id\n",
      "\t\t\tr_id = Rb_e[c_id][x];\n",
      "\t\t\tm_id = Ribosome[r_id].mRNA;\n",
      "\t\t\n",
      "\t\t\tif(Ribosome[r_id].pos>0) // If the ribosome is at the 2nd codon or higher (Ribosomes at 1st codon have empty P-sites).\n",
      "\t\t\t{\t\n",
      "\t\t\t\tdeacylatedFreeTrna[Ribosome[r_id].tRNA]++; // Move the tRNA at the P-site into the E-site and release it into deacylated free pool.\n",
      "\t\t\t\tribosomeBoundTrna[Ribosome[r_id].tRNA]--; \n",
      "\t\t\t}\n",
      "\n",
      "\t\t\tif (abort_flag == 0) // If aborting, no need to pick a tRNA from the acylated pool.\n",
      "\t\t\t\t{\n",
      "\t            // Pick a random A-site cognate tRNA based on respective elongation probabilities.\n",
      "\t            coin = rand()/((double)RAND_MAX);\n",
      "\t            prob_g = 0.0;\n",
      "\t    \n",
      "\t            for(cognatetrnaloop=0;cognatetrnaloop<2;cognatetrnaloop++)\n",
      "\t            {   \n",
      "\t                prob_g += ( Trna[ Codon[c_id].trna[cognatetrnaloop] ].trnaRibosomeRateConstant * acylatedFreeTrna[ Codon[c_id].trna[cognatetrnaloop] ] * Codon[c_id].weight[cognatetrnaloop] );\n",
      "\t    \n",
      "\t                if(coin<prob_g)\n",
      "\t                {   \n",
      "\t                    cognate_trna_id = Codon[c_id].trna[cognatetrnaloop]; \n",
      "\t                    break;\n",
      "\t                }   \n",
      "\t            }  \t\n",
      "\t\t\n",
      "\t\t\t\tRibosome[r_id].tRNA = cognate_trna_id; // Set the tRNA on the ribosome to be the chosen cognate tRNA.\n",
      "\t\t\t\tacylatedFreeTrna[Ribosome[r_id].tRNA]--; // Account for the tRNA that entered the ribosome from the free pool.\n",
      "\t\t\t\tribosomeBoundTrna[Ribosome[r_id].tRNA]++; // Account for the tRNA that entered the ribosome from the free pool.\n",
      "\t\t\t\t}\n",
      "\n",
      "\t\t\tif(currentTime > thresholdTime)\n",
      "\t\t\t\t{\t\n",
      "\t\t\t\tif(printOptions[0]==1)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tsummedElongationTimes[c_id] += currentTime-Ribosome[r_id].timeOfElongatableSwitch;\t\t\t\t// For estimation of avg elongation times of codons\n",
      "\t\t\t\t\tnumberOfElongations[c_id]++;\n",
      "\t\t\t\t\tif (abort_flag == 1)\n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\t\tnumberOfAbortions[c_id]++;\n",
      "\t\t\t\t\t\t}\t\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\tif(printOptions[4]==1)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tavgRibosomeDensityGrid[m_id][Ribosome[r_id].pos] += currentTime-Ribosome[r_id].timeOfLastElongation;\t\t// Keep track of the ribosome density at the corresponding codon position. \n",
      "\t\t\t\t\tnumberOfElongationsGrid[m_id][Ribosome[r_id].pos] ++;\t\t// Keep track of the number of times that this position on this mRNA was elongated. \n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\n",
      "\t\t\tif ( Ribosome[r_id].pos == (Gene[mRNA[m_id].gene].length-1) || (abort_flag == 1 ) )\t\t\t\t\t// Check if the current elongation has led to termination or abortion of translation.\n",
      "\t\t\t\t{\t\n",
      "\t\t\t\tR_grid[m_id][Ribosome[r_id].pos] = totalNumberOfRibosomes;\t\t\t\t\t\t// Update the position of ribosomes on the mRNA.\n",
      "\t\t\t\tif (abort_flag == 0) // If aborting, no trna was incorporated into the ribosome above.\n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\tdeacylatedFreeTrna[Ribosome[r_id].tRNA]++; // Release the ribosome bound tRNA back into the deacylated tRNA pool.\n",
      "\t\t\t\t\tribosomeBoundTrna[Ribosome[r_id].tRNA]--; // Release the ribosome bound tRNA back into the deacylated tRNA pool.\n",
      "\t\t\t\t\t}\n",
      "\n",
      "\t\t\t\tRibosome[r_id].tRNA = 0; // Reset the tRNA index on the ribosome. Note that this is swapped with the next_avail_ribosome value below.\n",
      "\t\t\t\tRibosome[r_id].pos++;\n",
      "\t\t\t\tRf++;\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t// Free a ribosome upon termination\n",
      "\n",
      "\t\t\t\tn_Rb_e[c_id]--;\n",
      "\t\t\t\tif(x!=n_Rb_e[c_id])\n",
      "\t\t\t\t{\tRibosome[Rb_e[c_id][n_Rb_e[c_id]]].elng_pos_list = x;\t\t\t// Update the ids and number of elongatable ribosomes\n",
      "\t\t\t\t\tRb_e[c_id][x] = Rb_e[c_id][n_Rb_e[c_id]];\n",
      "\t\t\t\t}\n",
      "\t\t\t\n",
      "\t\t\t\tif(currentTime>thresholdTime)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tif (abort_flag == 1)\n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tmRNA[m_id].numberOfAbortedProteins++;\n",
      "\t\t\t\t\t\tmRNA[m_id].averageAbortionTime += currentTime-Ribosome[r_id].timeOfLastInitiation;\t// Update the time to abortion\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\telse if (abort_flag == 0)\n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tmRNA[m_id].numberOfFullProteins++;\t\t\t\t\t\t\t\t\t\t// Update the number of trans evnts on curr mRNA\n",
      "\t\t\t\t\t\tmRNA[m_id].averageElongationTime += currentTime-Ribosome[r_id].timeOfLastInitiation;\t// Update the time to translation\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\n",
      "\t\t\t\t// Update any previously unelongatable ribosomes \n",
      "\t\t\t\tif(Ribosome[r_id].pos > 10 && R_grid[m_id][Ribosome[r_id].pos-11]<totalNumberOfRibosomes)\t\t\t\t\t\t// When the ribosome moves, a previously unelongatable ribosome\n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\tc2_id = Gene[mRNA[m_id].gene].seq[Ribosome[r_id].pos-11];\t\t\t// can now be elongatable on the same mRNA if its 11 codon behind\n",
      "\t\t\t\t\tRb_e[c2_id][n_Rb_e[c2_id]] = R_grid[m_id][Ribosome[r_id].pos-11];\n",
      "\t\t\t\t\t\n",
      "\t\t\t\t\tRibosome[R_grid[m_id][Ribosome[r_id].pos-11]].timeOfElongatableSwitch = currentTime;\n",
      "\t\t\t\t\tRibosome[R_grid[m_id][Ribosome[r_id].pos-11]].elng_cod_list = c2_id;\n",
      "\t\t\t\t\tRibosome[R_grid[m_id][Ribosome[r_id].pos-11]].elng_pos_list = n_Rb_e[c2_id];\n",
      "\t\t\t\t\tn_Rb_e[c2_id]++;\n",
      "\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t\t\t\t\t\n",
      "\t\t\t\tif ( Ribosome[r_id].pos <= 10 )\t\t\t\t\t\t\t\t\t\t\t\t\t// If the current position is less than or equal to codon 11, make the current mRNA initiable\n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\tg_id = mRNA[m_id].gene;\n",
      "\t\t\t\t\tfreeMrnaList[g_id][freeMrna[g_id]] = m_id;\n",
      "\t\t\t\t\tfreeMrna[g_id]++;\n",
      "\t\t\t\t\ttotalFreeMrna++;\n",
      "\t\t\t\t\tinitiationScaledMrnaCopyNumber[g_id] += Gene[g_id].ini_prob;\n",
      "\t\t\t\t\ttotalInitiationScaledMrnaCopyNumber += Gene[g_id].ini_prob;\n",
      "\t\t\t\t\t}\n",
      "\n",
      "\t\t\t\t// Update the ribosomes avail for initiation\n",
      "\t\t\t\tnext_avail_ribo--;\n",
      "\t\t\t\t\n",
      "\t\t\t\tif(r_id!=next_avail_ribo)\t\t\t\t\t\t\t\t\t\t\t// The last initiated ribosome's data is swapped with\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tRibosome[r_id].mRNA = Ribosome[next_avail_ribo].mRNA;\t\t\t\t\t// the ribosome that just finished translation (terminated)\n",
      "\t\t\t\t\tRibosome[r_id].pos = Ribosome[next_avail_ribo].pos;\t\t\t\t\t\t// This is done to primarily keep track of ONLY ribosomes\n",
      "\n",
      "\t\t\t\t\tRibosome[r_id].tRNA = Ribosome[next_avail_ribo].tRNA;\t\n",
      "\n",
      "\t\t\t\t\tRibosome[r_id].timeOfLastInitiation = Ribosome[next_avail_ribo].timeOfLastInitiation;\t\t// that are currently bound for faster computation\n",
      "\n",
      "\t\t\t\t\tRibosome[r_id].timeOfLastElongation = Ribosome[next_avail_ribo].timeOfLastElongation;\n",
      "\t\t\t\t\tRibosome[r_id].timeOfElongatableSwitch = Ribosome[next_avail_ribo].timeOfElongatableSwitch;\n",
      "\t\t\t\t\n",
      "\t\t\t\t\tRibosome[r_id].elng_cod_list = Gene[mRNA[Ribosome[next_avail_ribo].mRNA].gene].seq[Ribosome[next_avail_ribo].pos];\n",
      "\t\t\t\t\tRibosome[r_id].elng_pos_list = Ribosome[next_avail_ribo].elng_pos_list;\n",
      "\t\t\t\t\tR_grid[Ribosome[r_id].mRNA][Ribosome[r_id].pos] = r_id;\n",
      "\t\t\t\t\n",
      "\t\t\t\t\tif(R_grid[Ribosome[r_id].mRNA][Ribosome[r_id].pos+10]==totalNumberOfRibosomes || (Ribosome[r_id].pos+10)>=Gene[mRNA[Ribosome[r_id].mRNA].gene].length)\n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tRb_e[Ribosome[r_id].elng_cod_list][Ribosome[r_id].elng_pos_list] = r_id;\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\ttermtn_now = 1;\n",
      "\t\t\t\t}\n",
      "\n",
      "\t\t\telse if((Ribosome[r_id].pos+11)>=Gene[mRNA[m_id].gene].length || R_grid[m_id][Ribosome[r_id].pos+11]==totalNumberOfRibosomes)\t// Check if the ribosome is still elongatable\n",
      "\t\t\t{\tR_grid[m_id][Ribosome[r_id].pos] = totalNumberOfRibosomes;\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t// Update the position of ribosomes on the mRNA\n",
      "\t\t\t\t\n",
      "\t\t\t\tRibosome[r_id].pos++;\n",
      "\t\t\t\tR_grid[m_id][Ribosome[r_id].pos] = r_id;\n",
      "\t\t\t\tc2_id = Gene[mRNA[m_id].gene].seq[Ribosome[r_id].pos];\n",
      "\t\t\t\t\t\n",
      "\t\t\t\tif(c2_id!=c_id)\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t// If the codon has changed shift the elongatable ribosome\n",
      "\t\t\t\t{\tRb_e[c2_id][n_Rb_e[c2_id]] = r_id;\t\t\t\t\t\t\t\t\t// to the other codon\n",
      "\t\t\t\t\tRibosome[r_id].elng_cod_list = c2_id;\n",
      "\t\t\t\t\tRibosome[r_id].elng_pos_list = n_Rb_e[c2_id];\n",
      "\t\t\t\t\tn_Rb_e[c2_id]++;\n",
      "\t\t\t\t\t\n",
      "\t\t\t\t\tn_Rb_e[c_id]--;\n",
      "\t\t\t\t\tif(x!=n_Rb_e[c_id])\n",
      "\t\t\t\t\t{\tRibosome[Rb_e[c_id][n_Rb_e[c_id]]].elng_pos_list = x;\t\t\t\t// Update the ids and number of elongatable ribosomes\n",
      "\t\t\t\t\t\tRb_e[c_id][x] = Rb_e[c_id][n_Rb_e[c_id]];\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t}\n",
      "\t\t\t\n",
      "\t\t\telse\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t// When ribosome is not elongatable anymore\n",
      "\t\t\t{\tR_grid[m_id][Ribosome[r_id].pos] = totalNumberOfRibosomes;\t\t\t\t\t\t\t\t// Update the position of ribosomes on the mRNA\n",
      "\t\t\t\t\n",
      "\t\t\t\tRibosome[r_id].pos++;\n",
      "\t\t\t\tR_grid[m_id][Ribosome[r_id].pos] = r_id;\n",
      "\t\t\t\tRibosome[r_id].elng_cod_list = Gene[mRNA[m_id].gene].seq[Ribosome[r_id].pos];\n",
      "\n",
      "\t\t\t\tn_Rb_e[c_id]--;\n",
      "\t\t\t\tif(x!=n_Rb_e[c_id])\n",
      "\t\t\t\t{\tRibosome[Rb_e[c_id][n_Rb_e[c_id]]].elng_pos_list = x;\t\t\t\t\t// Update the ids and number of elongatable ribosomes\n",
      "\t\t\t\t\tRb_e[c_id][x] = Rb_e[c_id][n_Rb_e[c_id]];\n",
      "\t\t\t\t}\n",
      "\t\t\t\t\n",
      "\t\t\t}\n",
      "\t\t\t\n",
      "\t\t\tif(termtn_now==0)\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t// When elongation does not lead to termination\n",
      "\t\t\t{\tif(Ribosome[r_id].pos>10 && R_grid[m_id][Ribosome[r_id].pos-11]<totalNumberOfRibosomes)\t\t// When the ribosome moves, a previously unelongatable ribosome\n",
      "\t\t\t\t{\tc2_id = Gene[mRNA[m_id].gene].seq[Ribosome[r_id].pos-11];\t\t\t\t// can now be elongatable on the same mRNA if its 11 codon behind\n",
      "\t\t\t\t\tRb_e[c2_id][n_Rb_e[c2_id]] = R_grid[m_id][Ribosome[r_id].pos-11];\n",
      "\t\t\t\t\t\n",
      "\t\t\t\t\tRibosome[R_grid[m_id][Ribosome[r_id].pos-11]].timeOfElongatableSwitch = currentTime;\n",
      "\t\t\t\t\tRibosome[R_grid[m_id][Ribosome[r_id].pos-11]].elng_cod_list = c2_id;\n",
      "\t\t\t\t\tRibosome[R_grid[m_id][Ribosome[r_id].pos-11]].elng_pos_list = n_Rb_e[c2_id];\n",
      "\t\t\t\t\tn_Rb_e[c2_id]++;\n",
      "\t\t\t\t\t\n",
      "\t\t\t\t}\n",
      "\t\t\t\tif(Ribosome[r_id].pos==10)\t\t\t\t\t\t\t\t\t\t\t\t\t// If the current position is at codon 11 make the current mRNA initiable\n",
      "\t\t\t\t{\tg_id = mRNA[m_id].gene;\n",
      "\t\t\t\t\tfreeMrnaList[g_id][freeMrna[g_id]] = m_id;\n",
      "\t\t\t\t\tfreeMrna[g_id]++;\n",
      "\t\t\t\t\ttotalFreeMrna++;\n",
      "\t\t\t\t\tinitiationScaledMrnaCopyNumber[g_id] += Gene[g_id].ini_prob;\n",
      "\t\t\t\t\ttotalInitiationScaledMrnaCopyNumber += Gene[g_id].ini_prob;\n",
      "\t\t\t\t}\n",
      "\n",
      "\t\t\t\tRibosome[r_id].timeOfLastElongation = currentTime;\t\t\t\t\t\t\t\t\t\t\t\t// Upon elongation, update the elong ini time for the next evnt\n",
      "\t\t\t\tRibosome[r_id].timeOfElongatableSwitch = currentTime;\t\t\t\t\t\t\t\t\t\t\t\t// Upon elongation, update the elong ini time for the next evnt\n",
      "\t\t\t}\n",
      "\t\t\telse\n",
      "\t\t\t{\ttermtn_now = 0;\n",
      "\t\t\t}\n",
      "\t\t\t\n",
      "\t\t\tif(currentTime>((double)t_print) && printOptions[6]==1)\n",
      "\t\t\t{\tc3 = 0;\n",
      "\t\t\t\tstrcpy(outfile,outfilePrefix);\n",
      "\t\t\t\tsprintf(tmp_t,\"%d\",t_print);\n",
      "\t\t\t\tf1 = fopen(strcat(strcat(outfile,\"_ribo_pos_\"),tmp_t),\"w\");\t\t\t// Print ribosome pos every second\n",
      "\t\t\t\tprintf(\"%s\",outfile);\t\n",
      "\t\t\t\tfor(c1=0;c1<totalNumberOfGenes;c1++)\n",
      "\t\t\t\t{\tfor(c4=0;c4<Gene[c1].exp;c4++)\n",
      "\t\t\t\t\t{\tfor(c2=0;c2<Gene[c1].length;c2++)\n",
      "\t\t\t\t\t\t{\tif(R_grid[c3][c2]==totalNumberOfRibosomes)\n",
      "\t\t\t\t\t\t\t{\tfprintf(f1,\"0 \");\n",
      "\t\t\t\t\t\t\t}\n",
      "\t\t\t\t\t\t\telse\n",
      "\t\t\t\t\t\t\t{\tfprintf(f1,\"1 \");\n",
      "\t\t\t\t\t\t\t}\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t\tc3++;\n",
      "\t\t\t\t\t\tfprintf(f1,\"\\n\");\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t\tfclose(f1);\n",
      "\t\t\t\tt_print++;\n",
      "\t\t\t}\n",
      "\t\t}\n",
      "\tif (number_of_iterations%10000000 == 0)\n",
      "\t\t{\n",
      "\t\tprintf(\"iterations = %d\\ntime = %0.2fs\\ntotalRate = %0.2fs-1\\ntotalInitiationRate = %0.2fs-1\\ntotalElongationRate = %0.2fs-1\\ntotalAminoAcylationRate = %0.2fs-1\\n///////////////////////////////////////////\\n\", number_of_iterations, currentTime, totalRate, totalInitiationRate, totalElongationRate, totalAminoacylationRate);\n",
      "\t\t}\n",
      "\tnumber_of_iterations++;\n",
      "\n",
      "\t}\n",
      "\n",
      "\tfree(R_grid); // Free up the memory occupied by this huge matrix.\n",
      "\n",
      "//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////\n",
      "\t\n",
      "\t// Process output for printing\n",
      "\t\n",
      "\tint n_trans[totalNumberOfGenes];\n",
      "\tint n_aborts[totalNumberOfGenes];\n",
      "\tint n_tot=0;\n",
      "\tint n_abort_total=0;\n",
      "\tint n_tot_all=0;\n",
      "\tdouble g_etimes[totalNumberOfGenes];\n",
      "\tdouble g_atimes[totalNumberOfGenes];\n",
      "\tdouble g_ini[totalNumberOfGenes];\n",
      "\tdouble dCST=0.0;\n",
      "\t\n",
      "\tif(printOptions[1]==1 || printOptions[2]==1 || printOptions[3]==1)\n",
      "\t{\tc3 = 0;\n",
      "\t\tfor(c1=0;c1<totalNumberOfGenes;c1++)\n",
      "\t\t{\tn_trans[c1] = 0;\n",
      "\t\t\tg_ini[c1]=0.0;\n",
      "\t\t\tg_etimes[c1]=0.0;\n",
      "\t\t\t\n",
      "\t\t\tfor(c2=0;c2<Gene[c1].exp;c2++)\n",
      "\t\t\t\t{\t\n",
      "\t\t\t\tn_trans[c1] += mRNA[c3].numberOfFullProteins;\n",
      "\t\t\t\tn_aborts[c1] += mRNA[c3].numberOfAbortedProteins;\n",
      "\t\t\t\tg_ini[c1] += mRNA[c3].averageInitiationTime;\n",
      "\t\t\t\tg_etimes[c1] += mRNA[c3].averageElongationTime;\n",
      "\t\t\t\tg_atimes[c1] += mRNA[c3].averageAbortionTime;\n",
      "\t\t\t\tn_tot += mRNA[c3].numberOfFullProteins;\n",
      "\t\t\t\tn_abort_total += mRNA[c3].numberOfAbortedProteins;\n",
      "\t\t\t\tc3++;\n",
      "\t\t\t\t}\n",
      "\t\t}\n",
      "\t}\n",
      "\n",
      "\t\n",
      "\t// Print the output\n",
      "\n",
      "\t// Elongation times of all codons\n",
      "\tif(printOptions[0]==1)\n",
      "\t{\tstrcpy(outfile,outfilePrefix);\n",
      "\t\tf2 = fopen(strcat(outfile,\"_etimes.out\"),\"w\");\n",
      "\t\n",
      "\t\tfprintf(f2,\"Codon\\tNum_of_events\\tAvg_elong_time(sec)\\tNum_of_abortions\\n\");\n",
      "\t\tfor(c1=0;c1<61;c1++)\n",
      "\t\t\t{\t\n",
      "\t\t\tdCST = summedElongationTimes[c1]/(double)numberOfElongations[c1];\n",
      "\t\t\tfprintf(f2,\"%d\\t%s\\t%d\\t%0.4f\\t%d\\n\",c1, Codon[c1].triplet,numberOfElongations[c1],dCST, numberOfAbortions[c1] );\n",
      "\t\t\t}\n",
      "\t\tfclose(f2);\n",
      "\t}\n",
      "\t\n",
      "\t// Average total elongation times of all genes\n",
      "\tif(printOptions[1]==1)\n",
      "\t{\tstrcpy(outfile,outfilePrefix);\n",
      "\t\tf3 = fopen(strcat(outfile,\"_gene_totetimes.out\"),\"w\");\n",
      "\t\t\n",
      "\t\tfprintf(f3,\"Gene\\tNum_of_full_events\\tAvg_total_elong_time(sec)\\tNum_of_aborts\\tAvg_total_abort_time(sec)\\n\");\n",
      "\t\tfor(c1=0;c1<totalNumberOfGenes;c1++)\n",
      "\t\t{\tdCST = g_etimes[c1]/(double)n_trans[c1];\n",
      "\t\t\tfprintf(f3,\"%d\\t%d\\t%g\\t%d\\t%g\\n\",c1,n_trans[c1],dCST, n_aborts[c1], g_atimes[c1]/(double)n_aborts[c1]);\n",
      "\t\t}\n",
      "\t\tfclose(f3);\n",
      "\t}\n",
      "\t\n",
      "\t// Average time between initiation of all genes\n",
      "\tif(printOptions[2]==1)\n",
      "\t{\tstrcpy(outfile,outfilePrefix);\n",
      "\t\tf4 = fopen(strcat(outfile,\"_gene_initimes.out\"),\"w\");\n",
      "\t\t\n",
      "\t\tfprintf(f4,\"Gene\\tNum_of_events\\tAvg_initiation_time(sec)\\n\");\n",
      "\t\tfor(c1=0;c1<totalNumberOfGenes;c1++)\n",
      "\t\t{\tdCST = g_ini[c1]/(double)n_trans[c1];\n",
      "\t\t\tfprintf(f4,\"%d\\t%d\\t%g\\n\",c1,n_trans[c1],dCST);\n",
      "\t\t}\n",
      "\t\tfclose(f4);\n",
      "\t}\n",
      "\t\n",
      "\t// Average number of free ribosomes and tRNAs at equilibrium\n",
      "\tif(printOptions[3]==1)\n",
      "\t\t{\t\n",
      "\t\tstrcpy(outfile,outfilePrefix);\n",
      "\t\tf5 = fopen(strcat(outfile,\"_avg_ribo_tRNA.out\"),\"w\");\n",
      "\t\tfprintf( f5, \"Fractions of ribosomes (free/bound) and tRNAs (freeAcylated/freeDeacylated/ribosomeBound), Acylated Fraction\\n\" );\n",
      "\t\tavg_Rf = avg_Rf/(totalTime-thresholdTime)/totalNumberOfRibosomes;\n",
      "\t\tfprintf(f5,\"Ribosome\\t%0.6f\\t%0.6f\\n\",avg_Rf,1-avg_Rf);\n",
      "\t\tdouble acylatedFraction;\n",
      "\t\tfor(trnaloop=0;trnaloop<61;trnaloop++)\n",
      "\t\t\t{\n",
      "\t\t\tif (totalTrna[trnaloop] > 0)\n",
      "\t\t\t\t{\n",
      "\t\t\t\tavgAcylatedTrnaConcn[trnaloop] = avgAcylatedTrnaConcn[trnaloop]/(totalTime-thresholdTime)/totalTrna[trnaloop];\n",
      "\t\t\t\tavgDeacylatedTrnaConcn[trnaloop] = avgDeacylatedTrnaConcn[trnaloop]/(totalTime-thresholdTime)/totalTrna[trnaloop];\n",
      "\t\t\t\tavgRibosomeBoundTrnaConcn[trnaloop] = avgRibosomeBoundTrnaConcn[trnaloop]/(totalTime-thresholdTime)/totalTrna[trnaloop];\n",
      "\t\t\t\tacylatedFraction = avgAcylatedTrnaConcn[trnaloop] / (avgAcylatedTrnaConcn[trnaloop] + avgDeacylatedTrnaConcn[trnaloop]);\n",
      "\t\t\t\tfprintf(f5,\"tRNA%d\\t%s\\t%0.6f\\t%0.6f\\t%0.6f\\t%0.6f\\n\", trnaloop, Trna[trnaloop].anticodon, avgAcylatedTrnaConcn[trnaloop], avgDeacylatedTrnaConcn[trnaloop], avgRibosomeBoundTrnaConcn[trnaloop], acylatedFraction );\n",
      "\t\t\t\t}\n",
      "\t\t\t}\n",
      "\t\tfclose(f5);\n",
      "\t\t}\n",
      "\t\n",
      "\t// Time-averaged ribosome density on each gene that was tracked after threshold time was crossed.\n",
      "\tif(printOptions[4]==1)\n",
      "\t\t{\t\n",
      "\t\tstrcpy(outfile,outfilePrefix);\n",
      "\t\tf6 = fopen(strcat(outfile,\"_gene_ribo_density.out\"),\"w\");\n",
      "\t\tstrcpy(outfile,outfilePrefix);\n",
      "\t\tf6a = fopen(strcat(outfile,\"_gene_number_elongs.out\"),\"w\");\n",
      "\t\t\n",
      "\t\tfloat** geneSteadyStateRibosomeDensity;\t// Sum up the ribosome density on all mRNAs for each gene. \n",
      "\t\tint** geneTotalNumberOfElongations;\t// Sum up the total number of elongations on all mRNAs for each gene.\n",
      "\t\tgeneSteadyStateRibosomeDensity = (float**) malloc(sizeof(float *) * totalNumberOfGenes);\n",
      "\t\tgeneTotalNumberOfElongations = (int**) malloc(sizeof(int *) * totalNumberOfGenes);\n",
      "\n",
      "\t\tm_id = 0;\n",
      "\t\tint start_mid;\n",
      "\t\tfor(geneloop=0;geneloop<totalNumberOfGenes;geneloop++)\n",
      "\t\t\t{\t\n",
      "\t\t\tgeneSteadyStateRibosomeDensity[geneloop] = (float*) malloc(sizeof(float *) * obs_max_len);\n",
      "\t\t\tgeneTotalNumberOfElongations[geneloop] = (int*) malloc(sizeof(int *) * obs_max_len);\n",
      "\t\t\tstart_mid = m_id;\n",
      "\t\t\tfor(c1=0; c1<Gene[geneloop].length; c1++)\n",
      "\t\t\t\t{\n",
      "\t\t\t\tgeneSteadyStateRibosomeDensity[geneloop][c1] = 0.0;\n",
      "\t\t\t\tgeneTotalNumberOfElongations[geneloop][c1] = 0;\n",
      "\t\t\t\tfor (c2=0; c2<Gene[geneloop].exp; c2++)\n",
      "\t\t\t\t\t{\t\n",
      "\t\t\t\t\tgeneSteadyStateRibosomeDensity[geneloop][c1] += avgRibosomeDensityGrid[start_mid+c2][c1];\n",
      "\t\t\t\t\tgeneTotalNumberOfElongations[geneloop][c1] += numberOfElongationsGrid[start_mid+c2][c1];\n",
      "\t\t\t\t\tif (c1 == 0)\n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tm_id++;\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t}\n",
      "\n",
      "\t\tfor(geneloop=0; geneloop<totalNumberOfGenes; geneloop++)\n",
      "\t\t\t{\t\n",
      "\t\t\tfor(c2=0;c2<Gene[geneloop].length;c2++)\n",
      "\t\t\t\t{\t\n",
      "\t\t\t\tfprintf(f6,\"%0.2f \", geneSteadyStateRibosomeDensity[geneloop][c2]);\n",
      "\t\t\t\tfprintf(f6a,\"%d \", geneTotalNumberOfElongations[geneloop][c2]);\n",
      "\t\t\t\t}\n",
      "\t\t\tfprintf(f6,\"\\n\");\n",
      "\t\t\tfprintf(f6a,\"\\n\");\n",
      "\t\t\t}\n",
      "\t\tfclose(f6);\n",
      "\t\tfclose(f6a);\n",
      "\t\tfree( geneSteadyStateRibosomeDensity );\n",
      "\t\tfree( geneTotalNumberOfElongations );\n",
      "\t\t}\n",
      "\n",
      "\tFILE *fh;\t\n",
      "\tstrcpy( outfile, outfilePrefix);\n",
      "\tfh = fopen( strcat( outfile, \"_args.in\" ), \"a\" );\n",
      "\tgettimeofday(&time2, NULL);\n",
      "\n",
      "\tfprintf(fh, \"\\nProgram run time: %d seconds\\n\", time2.tv_sec - time1.tv_sec );\n",
      "    time_t result = time(NULL);\n",
      "\tfprintf(fh, \"Program end time: %s\", ctime(&result) );\n",
      "\tfclose(fh);\n",
      "\n",
      "    strcpy( outfile, outfilePrefix);\n",
      "    fh = fopen( strcat( outfile, \"_gene_data.in\" ), \"w\" );\n",
      "\tfprintf(fh, \"gene_id\\tlength\\texpression\\tinit_rate\\tfirst_10_codons\\n\" );\n",
      "\tfor (c1=0; c1<totalNumberOfGenes; c1++)\n",
      "\t\t{\n",
      "\t\tfprintf(fh, \"%d\\t%d\\t%d\\t%0.4f\", c1, Gene[c1].length, Gene[c1].exp, Gene[c1].ini_prob );\n",
      "\t\tif (Gene[c1].length >= 10)\n",
      "\t\t\t{\n",
      "\t\t\tfor (c2=0; c2<10; c2++)\n",
      "\t\t\t\t{\n",
      "\t\t\t\tfprintf( fh, \"\\t%d\", Gene[c1].seq[c2] );\n",
      "\t\t\t\t}\n",
      "\t\t\tfprintf( fh, \"\\n\" );\n",
      "\t\t\t}\n",
      "\t\t}\n",
      "    fclose(fh);\n",
      "\n",
      "    strcpy( outfile, outfilePrefix);\n",
      "    fh = fopen( strcat( outfile, \"_trna_data.in\" ), \"w\" );\n",
      "\tfprintf(fh, \"trna_id\\tanticodon\\tconcn\\tribosome_association_rate\\taminoacylation_rate\\n\" );\n",
      "\tfor (c1=0; c1<61; c1++)\n",
      "\t\t{\n",
      "\t\tif (Trna[c1].concn > 0)\n",
      "\t\t\t{\n",
      "\t\t\tfprintf(fh, \"%d\\t%s\\t%0.0f\\t%0.3g\\t%0.3g\\n\", c1, Trna[c1].anticodon, \n",
      "\t\t\t\tTrna[c1].concn, Trna[c1].trnaRibosomeRateConstant, Trna[c1].aminoacylationRateConstant );\n",
      "\t\t\t}\n",
      "\t\t}\n",
      "    fclose(fh);\n",
      "\n",
      "    strcpy( outfile, outfilePrefix);\n",
      "    fh = fopen( strcat( outfile, \"_codon_data.in\" ), \"w\" );\n",
      "\tfprintf(fh, \"codon_id\\tcodon\\ttrna1\\ttrna2\\tweight1\\tweight2\\n\" );\n",
      "\tfor (c1=0; c1<61; c1++)\n",
      "\t\t{\n",
      "\t\tfprintf(fh, \"%d\\t%s\\t%0d\\t%d\\t%0.2f\\t%0.2f\\n\", c1, Codon[c1].triplet,\n",
      "\t\t\t\t\tCodon[c1].trna[0], Codon[c1].trna[1], Codon[c1].weight[0], Codon[c1].weight[1] ); \n",
      "\t\t}\n",
      "    fclose(fh);\n",
      "\t\t\n",
      "    strcpy( outfile, outfilePrefix);\n",
      "    fh = fopen( strcat( outfile, \"_global_vars_data.in\" ), \"w\" );\n",
      "\tfprintf(fh, \"Random Seed = %d (for RNG initialization)\\nCodon Independent Elongation Time = %0.3f sec\\nNumber of Genes = %d\\nNumber of Ribosomes = %d per cell\\nBound Ribosome Fraction = %0.3f\\nNumber of tRNA = %d per cell\\nNumber of mRNA = %d per cell\\nCell Volume = %0.3g m-3\\nTotal Simulation Time = %0.0f sec\\nThreshold Time = %0.0f sec\\nRibosome-tRNA kcat / Km = %0.3g M-1s-1\\n\", seed, codonMinElongationTime, totalNumberOfGenes, totalNumberOfRibosomes, boundRibosomeFraction, totalNumberOfTrna,totalNumberOfMrna, cellVolume, totalTime, thresholdTime, ribosomeTrnaKcatByKm);\n",
      "\n",
      "\t// Free the malloc structures and arrays\n",
      "\tfree(Gene);\n",
      "\tfree(Ribosome);\n",
      "\tfree(mRNA);\n",
      "\tfree(Codon);\n",
      "\tfree(Trna);\n",
      "\tfree(Rb_e);\n",
      "\tfree(avgRibosomeDensityGrid);\n",
      "\tfree(numberOfElongationsGrid);\n",
      "\n",
      "\tprintf(\"Finished Simulation. Total time: %0.2fs.\\n\",currentTime);\n",
      "}\n",
      "```"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<a id=\"compile\"></a>\n",
      "### Compiling source  ###"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash \n",
      "\n",
      "gcc scripts/ecoli_translation.c -w -lm -o scripts/ecoli_translation\n",
      "\n",
      "# used gcc version 4.8.2\n",
      "# -w suppresses warnings.\n",
      "# -lm links math library.\n",
      "# -o specifies output filename."
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<a id=\"run\"></a>\n",
      "### Running the simulation  ###"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash \n",
      "\n",
      "# Create directories for holding the output files from the simulation\n",
      "mkdir output_files output_files/simulation_v1 output_files/simulation_v2 output_files/simulation_v3 output_files/simulation_v4 output_files/simulation_v5 output_files/simulation_v6"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Parameter sweeps for different figures"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%writefile scripts/run_simulation.py\n",
      "\n",
      "import sys\n",
      "import subprocess\n",
      "import numpy\n",
      "\n",
      "jobindex = 0\n",
      "\n",
      "# Simulations with differential aminoacylation rates for leucine and serine trna and nonzero time for intraribosomal events.\n",
      "# (Figures 3,5,7)\n",
      "for (aminoacid, relativetrnaaminoacylationrate) in zip( ['leu','ser'], ['0.5 1.5 0.5 0.5 1', '0.5 1.5 1 0.5'] ):\n",
      "    for (ribotrnakcatkm, codonminelongtime) in zip( [\"2e7\"], [\"0.05\"] ):\n",
      "        for ribosomes in [44000]:\n",
      "            for starvationlevel in [0.01] + list( (2.0)**(numpy.arange(0,-10,-0.5))):\n",
      "                for abortionrate in [0.00, 0.1] + list((2.0)**(numpy.arange(5.0,-10.0,-1.0))):\n",
      "                    jobindex += 1\n",
      "                    if not jobindex == int(sys.argv[1]):\n",
      "                        continue\n",
      "                    subprocess.call( [ \"./scripts/ecoli_translation\" + \n",
      "                                      \" --threshold-time 1000\" + \n",
      "                                      \" --total-time 1500\" +\n",
      "                                      \" --aminoacid \" + aminoacid  + \" starvationfactor \" + \"%0.6f\"%starvationlevel +\n",
      "                                      \" --total-ribosomes \" + str(ribosomes) + \n",
      "                                      \" --aminoacid \" + aminoacid + \" relativetrnaaminoacylationrates \" + relativetrnaaminoacylationrate + \n",
      "                                      \" --abort-threshold-rate 10\" + \n",
      "                                      \" --abortion-relative-rate \" + \"%0.6f\"%abortionrate +\n",
      "                                      \" --ribotrnakcatkm \" + ribotrnakcatkm +\n",
      "                                      \" --min-elong-time \" + codonminelongtime + \n",
      "                                      \" --output-prefix output_files/simulation_v1/output_v1\" +  \n",
      "                                      \" --input-genes seq_and_annotation/Ecoli.input.seq\"  ],\n",
      "                                    shell= True )\n",
      "                    \n",
      "print jobindex                    \n",
      "                    \n",
      "# Simulations to compare nonzero and zero time for intraribosomal events (Figure 3)\n",
      "for (ribotrnakcatkm, codonminelongtime) in zip( [\"2e7\",\"2e6\"], [\"0.05\", \"0.00\"] ):\n",
      "    for ribosomes in [44000]:\n",
      "        for abortionrate in [0.1]:\n",
      "            jobindex += 1\n",
      "            if not jobindex == int(sys.argv[1]):\n",
      "                continue\n",
      "            subprocess.call( [ \"./scripts/ecoli_translation\" + \n",
      "                              \" --threshold-time 1000\" + \n",
      "                              \" --total-time 1500\" +\n",
      "                              \" --total-ribosomes \" + str(ribosomes) + \n",
      "                              \" --abort-threshold-rate 10\" + \n",
      "                              \" --abortion-relative-rate \" + \"%0.6f\"%abortionrate +\n",
      "                              \" --ribotrnakcatkm \" + ribotrnakcatkm +\n",
      "                              \" --min-elong-time \" + codonminelongtime + \n",
      "                              \" --output-prefix output_files/simulation_v2/output_v2\" +  \n",
      "                              \" --input-genes seq_and_annotation/Ecoli.input.seq\"  ],\n",
      "                            shell= True )\n",
      "print jobindex                    \n",
      "\n",
      "# Simulations to overexpress transgenes (Figure 3)\n",
      "for (ribotrnakcatkm, codonminelongtime) in zip( [\"2e7\",\"2e6\"], [\"0.05\", \"0.00\"] ):\n",
      "    for ribosomes in [44000]:\n",
      "        for abortionrate in [0.1]:\n",
      "            for mrnaFraction in map( str, [1, 10, 50] ):\n",
      "                for cai in map( lambda x: \"%0.1f\"%x, numpy.arange(0.1, 1, 0.1) ):\n",
      "                    jobindex += 1\n",
      "                    if not jobindex == int(sys.argv[1]):\n",
      "                        continue\n",
      "                    subprocess.call( [ \"./scripts/ecoli_translation\" + \n",
      "                                      \" --threshold-time 1000\" + \n",
      "                                      \" --total-time 1500\" +\n",
      "                                      \" --total-ribosomes \" + str(ribosomes) + \n",
      "                                      \" --abort-threshold-rate 10\" + \n",
      "                                      \" --abortion-relative-rate \" + \"%0.6f\"%abortionrate +\n",
      "                                      \" --ribotrnakcatkm \" + ribotrnakcatkm +\n",
      "                                      \" --min-elong-time \" + codonminelongtime + \n",
      "                                      \" --output-prefix output_files/simulation_v3/output_v3_mrnafraction_\" + mrnaFraction + \"_cai_\" + cai + \n",
      "                                      \" --input-genes seq_and_annotation/Ecoli.input.transgenefrac\" + mrnaFraction + \".cai\" + cai + \".seq\"   ],\n",
      "                                    shell= True )\n",
      "print jobindex         \n",
      "\n",
      "# Simulations with equal aminoacylation rates for leucine and serine trna. (Figure 3)\n",
      "for (aminoacid, relativetrnaaminoacylationrate) in zip( ['leu','ser'], ['1 1 1 1 1','1 1 1 1'] ):\n",
      "    for (ribotrnakcatkm, codonminelongtime) in zip( [\"2e7\"], [\"0.05\"] ):\n",
      "        for ribosomes in [44000]:\n",
      "            for starvationlevel in list( (2.0)**(numpy.arange(0,-10,-0.5))):                \n",
      "                for abortionrate in [0.1]:\n",
      "                    jobindex += 1\n",
      "                    if not jobindex == int(sys.argv[1]):\n",
      "                        continue\n",
      "                    subprocess.call( [ \"./scripts/ecoli_translation\" + \n",
      "                                      \" --threshold-time 1000\" + \n",
      "                                      \" --total-time 1500\" +\n",
      "                                      \" --aminoacid \" + aminoacid  + \" starvationfactor \" + \"%0.6f\"%starvationlevel +\n",
      "                                      \" --total-ribosomes \" + str(ribosomes) + \n",
      "                                      \" --aminoacid \" + aminoacid + \" relativetrnaaminoacylationrates \" + relativetrnaaminoacylationrate + \n",
      "                                      \" --abort-threshold-rate 10\" + \n",
      "                                      \" --abortion-relative-rate \" + \"%0.6f\"%abortionrate +\n",
      "                                      \" --ribotrnakcatkm \" + ribotrnakcatkm +\n",
      "                                      \" --min-elong-time \" + codonminelongtime + \n",
      "                                      \" --output-prefix output_files/simulation_v4/output_v4\" +  \n",
      "                                      \" --input-genes seq_and_annotation/Ecoli.input.seq\"  ],\n",
      "                                    shell= True )\n",
      "print jobindex                    \n",
      "\n",
      "# Simulations of leucine starvation with codon-trna kcat/KM rates from Sorensen et al. J. Mol. Biol. 2005.\n",
      "# (Figure S2C). Copy ecoli_translation.c to ecoli_translation_sorensen.c with the input cognate weights file\n",
      "# changed in the code. Compile this as ecoli_translation_sorensen and then run the loop below.\n",
      "for (aminoacid, relativetrnaaminoacylationrate) in zip( ['leu'], ['1 1 1 1 1'] ):\n",
      "    for (ribotrnakcatkm, codonminelongtime) in zip( [\"2e7\"], [\"0.05\"] ):\n",
      "        for ribosomes in [44000]:\n",
      "            for starvationlevel in [0.01] + list( (2.0)**(numpy.arange(0,-10,-0.5))):\n",
      "                for abortionrate in [0.1]:\n",
      "                    jobindex += 1\n",
      "                    if not jobindex == int(sys.argv[1]):\n",
      "                        continue\n",
      "                    subprocess.call( [ \"./scripts/ecoli_translation_sorensen\" + \n",
      "                                      \" --threshold-time 1000\" + \n",
      "                                      \" --total-time 1500\" +\n",
      "                                      \" --aminoacid \" + aminoacid  + \" starvationfactor \" + \"%0.6f\"%starvationlevel +\n",
      "                                      \" --total-ribosomes \" + str(ribosomes) + \n",
      "                                      \" --aminoacid \" + aminoacid + \" relativetrnaaminoacylationrates \" + relativetrnaaminoacylationrate + \n",
      "                                      \" --abort-threshold-rate 10\" + \n",
      "                                      \" --abortion-relative-rate \" + \"%0.6f\"%abortionrate +\n",
      "                                      \" --ribotrnakcatkm \" + ribotrnakcatkm +\n",
      "                                      \" --min-elong-time \" + codonminelongtime + \n",
      "                                      \" --output-prefix output_files/simulation_v5/output_v5\" +  \n",
      "                                      \" --input-genes seq_and_annotation/Ecoli.input.seq\"  ],\n",
      "                                    shell= True )\n",
      "print jobindex  \n",
      "\n",
      "#Simulation to vary the difference in aminoacylation rates between different tRNAs (Figure S2D)\n",
      "p = numpy.arange(0,1,0.1)\n",
      "relativeaminoacylationrates = []\n",
      "for row in numpy.round(zip(1-p,1+p,1-p,1-p,numpy.ones(10,)),2):\n",
      "    temp = ''\n",
      "    for col in row:\n",
      "        temp = temp + '%0.1f '%col\n",
      "    relativeaminoacylationrates.append( temp.strip() )\n",
      "\n",
      "for aminoacid in ['leu']:\n",
      "    for relativetrnaaminoacylationrate in relativeaminoacylationrates:\n",
      "        for (ribotrnakcatkm, codonminelongtime) in zip( [\"2e7\"], [\"0.05\"] ):\n",
      "            for ribosomes in [44000]:\n",
      "                for starvationlevel in [(2.0)**(-10)]:                \n",
      "                    for abortionrate in [0.1]:\n",
      "                        jobindex += 1\n",
      "                        if not jobindex == int(sys.argv[1]):\n",
      "                            continue\n",
      "                        subprocess.call( [ \"./scripts/ecoli_translation\" + \n",
      "                                          \" --threshold-time 1000\" + \n",
      "                                          \" --total-time 1500\" +\n",
      "                                          \" --aminoacid \" + aminoacid  + \" starvationfactor \" + \"%0.6f\"%starvationlevel +\n",
      "                                          \" --total-ribosomes \" + str(ribosomes) + \n",
      "                                          \" --aminoacid \" + aminoacid + \" relativetrnaaminoacylationrates \" + relativetrnaaminoacylationrate + \n",
      "                                          \" --abort-threshold-rate 10\" + \n",
      "                                          \" --abortion-relative-rate \" + \"%0.6f\"%abortionrate +\n",
      "                                          \" --ribotrnakcatkm \" + ribotrnakcatkm +\n",
      "                                          \" --min-elong-time \" + codonminelongtime + \n",
      "                                          \" --output-prefix output_files/simulation_v6/output_v6\" +  \n",
      "                                          \" --input-genes seq_and_annotation/Ecoli.input.seq\"  ],\n",
      "                                        shell= True )\n",
      "print jobindex "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Overwriting scripts/run_simulation.py\n"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Total of 841 simulatons for above set of parameters.\n",
      "\n",
      "Each simulation run takes ~3 hours on a 32-bit single core laptop.\n",
      "\n",
      "For fast run times, run each iteration on a separate node of a cluster. Alternately reduce the run time above from 1500s to 150s and threshold time from 1000s to 100s."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash\n",
      "\n",
      "for jobindex in {1..841};\n",
      "    do\n",
      "    python scripts/run_simulation.py $jobindex;\n",
      "    done;"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}